// @kubb/swagger-client/dist/chunk-HOJIV7MX.js
// src/components/Client.tsx
import { URLPath } from "@kubb/core/utils";
import { useOperation, useOperationManager } from "@kubb/plugin-oas/hooks";
import { getComments, getPathParams } from "@kubb/plugin-oas/utils";
import { File, Function, Parser, useApp } from "@kubb/react";
import { pluginTsName } from "@kubb/swagger-ts";
import { isOptional } from "@kubb/oas";
import { jsx, jsxs } from "@kubb/react/jsx-runtime";
function Template({ name, generics, returnType, params, JSDoc, client }) {
  const isFormData = client.contentType === "multipart/form-data";
  const headers = [
    client.contentType !== "application/json" ? `'Content-Type': '${client.contentType}'` : void 0,
    client.withHeaders ? "...headers" : void 0
  ].filter(Boolean).join(", ");
  const clientParams = {
    data: {
      mode: "object",
      children: {
        method: {
          type: "string",
          value: JSON.stringify(client.method)
        },
        url: {
          type: "string",
          value: client.path.template
        },
        baseURL: client.baseURL ? {
          type: "string",
          value: JSON.stringify(client.baseURL)
        } : void 0,
        params: client.withQueryParams ? {
          type: "any"
        } : void 0,
        data: client.withData ? {
          type: "any",
          value: isFormData ? "formData" : void 0
        } : void 0,
        headers: headers.length ? {
          type: "any",
          value: headers.length ? `{ ${headers}, ...options.headers }` : void 0
        } : void 0,
        options: {
          type: "any",
          mode: "inlineSpread"
        }
      }
    }
  };
  const formData = isFormData ? `
   const formData = new FormData()
   if(data) {
    Object.keys(data).forEach((key) => {
      const value = data[key];
      if (typeof key === "string" && (typeof value === "string" || value instanceof Blob)) {
        formData.append(key, value);
      }
    })
   }
  ` : void 0;
  return /* @__PURE__ */ jsxs(Function, { name, async: true, export: true, generics, returnType, params, JSDoc, children: [
    formData || "",
    /* @__PURE__ */ jsx(Function.Call, { name: "res", to: /* @__PURE__ */ jsx(Function, { name: "client", async: true, generics: client.generics, params: clientParams }) }),
    /* @__PURE__ */ jsx(Function.Return, { children: client.dataReturnType === "data" ? "res.data" : "res" })
  ] });
}
function RootTemplate({ children }) {
  const {
    plugin: {
      options: {
        client: { importPath },
        extName
      }
    }
  } = useApp();
  const { getSchemas, getFile } = useOperationManager();
  const operation = useOperation();
  const file = getFile(operation);
  const fileType = getFile(operation, { pluginKey: [pluginTsName] });
  const schemas = getSchemas(operation, { pluginKey: [pluginTsName], type: "type" });
  return /* @__PURE__ */ jsx(Parser, { language: "typescript", children: /* @__PURE__ */ jsxs(File, { baseName: file.baseName, path: file.path, meta: file.meta, children: [
    /* @__PURE__ */ jsx(File.Import, { name: "client", path: importPath }),
    /* @__PURE__ */ jsx(File.Import, { name: ["ResponseConfig"], path: importPath, isTypeOnly: true }),
    /* @__PURE__ */ jsx(
      File.Import,
      {
        extName,
        name: [schemas.request?.name, schemas.response.name, schemas.pathParams?.name, schemas.queryParams?.name, schemas.headerParams?.name].filter(Boolean),
        root: file.path,
        path: fileType.path,
        isTypeOnly: true
      }
    ),
    /* @__PURE__ */ jsx(File.Source, { children })
  ] }) });
}
var defaultTemplates = { default: Template, root: RootTemplate };
function Client({ baseURL, Template: Template3 = defaultTemplates.default }) {
  const {
    plugin: {
      options: { client, dataReturnType, pathParamsType }
    }
  } = useApp();
  const { getSchemas, getName } = useOperationManager();
  const operation = useOperation();
  const contentType = operation.getContentType();
  const name = getName(operation, { type: "function" });
  const schemas = getSchemas(operation, { pluginKey: [pluginTsName], type: "type" });
  return /* @__PURE__ */ jsx(
    Template3,
    {
      name,
      params: {
        pathParams: {
          mode: pathParamsType === "object" ? "object" : "inlineSpread",
          children: getPathParams(schemas.pathParams, { typed: true })
        },
        data: schemas.request?.name ? {
          type: schemas.request?.name,
          optional: isOptional(schemas.request?.schema)
        } : void 0,
        params: schemas.queryParams?.name ? {
          type: schemas.queryParams?.name,
          optional: isOptional(schemas.queryParams?.schema)
        } : void 0,
        headers: schemas.headerParams?.name ? {
          type: schemas.headerParams?.name,
          optional: isOptional(schemas.headerParams?.schema)
        } : void 0,
        options: {
          type: "Partial<Parameters<typeof client>[0]>",
          default: "{}"
        }
      },
      returnType: dataReturnType === "data" ? `ResponseConfig<${schemas.response.name}>["data"]` : `ResponseConfig<${schemas.response.name}>`,
      JSDoc: {
        comments: getComments(operation)
      },
      client: {
        baseURL,
        generics: [schemas.response.name, schemas.request?.name].filter(Boolean),
        dataReturnType,
        withQueryParams: !!schemas.queryParams?.name,
        withData: !!schemas.request?.name,
        withHeaders: !!schemas.headerParams?.name,
        method: operation.method,
        path: new URLPath(operation.path),
        contentType
      }
    }
  );
}
Client.File = function({ baseURL, ...props }) {
  const templates = { ...defaultTemplates, ...props.templates };
  const Template3 = templates.default;
  const RootTemplate3 = templates.root;
  return /* @__PURE__ */ jsx(RootTemplate3, { children: /* @__PURE__ */ jsx(Client, { baseURL, Template: Template3 }) });
};
Client.templates = defaultTemplates;

// src/components/Operations.tsx
import { URLPath as URLPath2 } from "@kubb/core/utils";
import { useOperations } from "@kubb/plugin-oas/hooks";
import { Const, File as File2, Parser as Parser2, useApp as useApp2 } from "@kubb/react";
import { jsx as jsx2 } from "@kubb/react/jsx-runtime";
function Template2({ name, operations }) {
  const operationsObject = {};
  operations.forEach((operation) => {
    operationsObject[operation.getOperationId()] = {
      path: new URLPath2(operation.path).URL,
      method: operation.method
    };
  });
  return /* @__PURE__ */ jsx2(Const, { name, export: true, asConst: true, children: JSON.stringify(operationsObject, void 0, 2) });
}
function RootTemplate2({ children }) {
  const {
    pluginManager,
    plugin: { key: pluginKey }
  } = useApp2();
  const file = pluginManager.getFile({ name: "operations", extName: ".ts", pluginKey });
  return /* @__PURE__ */ jsx2(Parser2, { language: "typescript", children: /* @__PURE__ */ jsx2(File2, { baseName: file.baseName, path: file.path, meta: file.meta, exportable: false, children: /* @__PURE__ */ jsx2(File2.Source, { children }) }) });
}
var defaultTemplates2 = { default: Template2, root: RootTemplate2 };
function Operations({ baseURL, Template: Template3 = defaultTemplates2.default }) {
  const operations = useOperations();
  return /* @__PURE__ */ jsx2(Template3, { baseURL, name: "operations", operations });
}
Operations.File = function({ baseURL, ...props }) {
  const templates = { ...defaultTemplates2, ...props.templates };
  const Template3 = templates.default;
  const RootTemplate3 = templates.root;
  return /* @__PURE__ */ jsx2(RootTemplate3, { children: /* @__PURE__ */ jsx2(Operations, { baseURL, Template: Template3 }) });
};
Operations.templates = defaultTemplates2;

export {
  Client,
  Operations
};
