// @kubb/swagger-client/dist/chunk-WGN7DKLR.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _optionalChain(ops) { let lastAccessLHS = undefined; let value = ops[0]; let i = 1; while (i < ops.length) { const op = ops[i]; const fn = ops[i + 1]; i += 2; if ((op === 'optionalAccess' || op === 'optionalCall') && value == null) { return undefined; } if (op === 'access' || op === 'optionalAccess') { lastAccessLHS = value; value = fn(value); } else if (op === 'call' || op === 'optionalCall') { value = fn((...args) => value.call(lastAccessLHS, ...args)); lastAccessLHS = undefined; } } return value; }// src/components/Client.tsx
var _utils = require('@kubb/core/utils');
var _hooks = require('@kubb/plugin-oas/hooks');
var _utils3 = require('@kubb/plugin-oas/utils');
var _react = require('@kubb/react');
var _swaggerts = require('@kubb/swagger-ts');
var _oas = require('@kubb/oas');
var _jsxruntime = require('@kubb/react/jsx-runtime');
function Template({ name, generics, returnType, params, JSDoc, client }) {
  const isFormData = client.contentType === "multipart/form-data";
  const headers = [
    client.contentType !== "application/json" ? `'Content-Type': '${client.contentType}'` : void 0,
    client.withHeaders ? "...headers" : void 0
  ].filter(Boolean).join(", ");
  const clientParams = {
    data: {
      mode: "object",
      children: {
        method: {
          type: "string",
          value: JSON.stringify(client.method)
        },
        url: {
          type: "string",
          value: client.path.template
        },
        baseURL: client.baseURL ? {
          type: "string",
          value: JSON.stringify(client.baseURL)
        } : void 0,
        params: client.withQueryParams ? {
          type: "any"
        } : void 0,
        data: client.withData ? {
          type: "any",
          value: isFormData ? "formData" : void 0
        } : void 0,
        headers: headers.length ? {
          type: "any",
          value: headers.length ? `{ ${headers}, ...options.headers }` : void 0
        } : void 0,
        options: {
          type: "any",
          mode: "inlineSpread"
        }
      }
    }
  };
  const formData = isFormData ? `
   const formData = new FormData()
   if(data) {
    Object.keys(data).forEach((key) => {
      const value = data[key];
      if (typeof key === "string" && (typeof value === "string" || value instanceof Blob)) {
        formData.append(key, value);
      }
    })
   }
  ` : void 0;
  return /* @__PURE__ */ _jsxruntime.jsxs.call(void 0, _react.Function, { name, async: true, export: true, generics, returnType, params, JSDoc, children: [
    formData || "",
    /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _react.Function.Call, { name: "res", to: /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _react.Function, { name: "client", async: true, generics: client.generics, params: clientParams }) }),
    /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _react.Function.Return, { children: client.dataReturnType === "data" ? "res.data" : "res" })
  ] });
}
function RootTemplate({ children }) {
  const {
    plugin: {
      options: {
        client: { importPath },
        extName
      }
    }
  } = _react.useApp.call(void 0, );
  const { getSchemas, getFile } = _hooks.useOperationManager.call(void 0, );
  const operation = _hooks.useOperation.call(void 0, );
  const file = getFile(operation);
  const fileType = getFile(operation, { pluginKey: [_swaggerts.pluginTsName] });
  const schemas = getSchemas(operation, { pluginKey: [_swaggerts.pluginTsName], type: "type" });
  return /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _react.Parser, { language: "typescript", children: /* @__PURE__ */ _jsxruntime.jsxs.call(void 0, _react.File, { baseName: file.baseName, path: file.path, meta: file.meta, children: [
    /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _react.File.Import, { name: "client", path: importPath }),
    /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _react.File.Import, { name: ["ResponseConfig"], path: importPath, isTypeOnly: true }),
    /* @__PURE__ */ _jsxruntime.jsx.call(void 0, 
      _react.File.Import,
      {
        extName,
        name: [_optionalChain([schemas, 'access', _ => _.request, 'optionalAccess', _2 => _2.name]), schemas.response.name, _optionalChain([schemas, 'access', _3 => _3.pathParams, 'optionalAccess', _4 => _4.name]), _optionalChain([schemas, 'access', _5 => _5.queryParams, 'optionalAccess', _6 => _6.name]), _optionalChain([schemas, 'access', _7 => _7.headerParams, 'optionalAccess', _8 => _8.name])].filter(Boolean),
        root: file.path,
        path: fileType.path,
        isTypeOnly: true
      }
    ),
    /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _react.File.Source, { children })
  ] }) });
}
var defaultTemplates = { default: Template, root: RootTemplate };
function Client({ baseURL, Template: Template3 = defaultTemplates.default }) {
  const {
    plugin: {
      options: { client, dataReturnType, pathParamsType }
    }
  } = _react.useApp.call(void 0, );
  const { getSchemas, getName } = _hooks.useOperationManager.call(void 0, );
  const operation = _hooks.useOperation.call(void 0, );
  const contentType = operation.getContentType();
  const name = getName(operation, { type: "function" });
  const schemas = getSchemas(operation, { pluginKey: [_swaggerts.pluginTsName], type: "type" });
  return /* @__PURE__ */ _jsxruntime.jsx.call(void 0, 
    Template3,
    {
      name,
      params: {
        pathParams: {
          mode: pathParamsType === "object" ? "object" : "inlineSpread",
          children: _utils3.getPathParams.call(void 0, schemas.pathParams, { typed: true })
        },
        data: _optionalChain([schemas, 'access', _9 => _9.request, 'optionalAccess', _10 => _10.name]) ? {
          type: _optionalChain([schemas, 'access', _11 => _11.request, 'optionalAccess', _12 => _12.name]),
          optional: _oas.isOptional.call(void 0, _optionalChain([schemas, 'access', _13 => _13.request, 'optionalAccess', _14 => _14.schema]))
        } : void 0,
        params: _optionalChain([schemas, 'access', _15 => _15.queryParams, 'optionalAccess', _16 => _16.name]) ? {
          type: _optionalChain([schemas, 'access', _17 => _17.queryParams, 'optionalAccess', _18 => _18.name]),
          optional: _oas.isOptional.call(void 0, _optionalChain([schemas, 'access', _19 => _19.queryParams, 'optionalAccess', _20 => _20.schema]))
        } : void 0,
        headers: _optionalChain([schemas, 'access', _21 => _21.headerParams, 'optionalAccess', _22 => _22.name]) ? {
          type: _optionalChain([schemas, 'access', _23 => _23.headerParams, 'optionalAccess', _24 => _24.name]),
          optional: _oas.isOptional.call(void 0, _optionalChain([schemas, 'access', _25 => _25.headerParams, 'optionalAccess', _26 => _26.schema]))
        } : void 0,
        options: {
          type: "Partial<Parameters<typeof client>[0]>",
          default: "{}"
        }
      },
      returnType: dataReturnType === "data" ? `ResponseConfig<${schemas.response.name}>["data"]` : `ResponseConfig<${schemas.response.name}>`,
      JSDoc: {
        comments: _utils3.getComments.call(void 0, operation)
      },
      client: {
        baseURL,
        generics: [schemas.response.name, _optionalChain([schemas, 'access', _27 => _27.request, 'optionalAccess', _28 => _28.name])].filter(Boolean),
        dataReturnType,
        withQueryParams: !!_optionalChain([schemas, 'access', _29 => _29.queryParams, 'optionalAccess', _30 => _30.name]),
        withData: !!_optionalChain([schemas, 'access', _31 => _31.request, 'optionalAccess', _32 => _32.name]),
        withHeaders: !!_optionalChain([schemas, 'access', _33 => _33.headerParams, 'optionalAccess', _34 => _34.name]),
        method: operation.method,
        path: new (0, _utils.URLPath)(operation.path),
        contentType
      }
    }
  );
}
Client.File = function({ baseURL, ...props }) {
  const templates = { ...defaultTemplates, ...props.templates };
  const Template3 = templates.default;
  const RootTemplate3 = templates.root;
  return /* @__PURE__ */ _jsxruntime.jsx.call(void 0, RootTemplate3, { children: /* @__PURE__ */ _jsxruntime.jsx.call(void 0, Client, { baseURL, Template: Template3 }) });
};
Client.templates = defaultTemplates;

// src/components/Operations.tsx




function Template2({ name, operations }) {
  const operationsObject = {};
  operations.forEach((operation) => {
    operationsObject[operation.getOperationId()] = {
      path: new (0, _utils.URLPath)(operation.path).URL,
      method: operation.method
    };
  });
  return /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _react.Const, { name, export: true, asConst: true, children: JSON.stringify(operationsObject, void 0, 2) });
}
function RootTemplate2({ children }) {
  const {
    pluginManager,
    plugin: { key: pluginKey }
  } = _react.useApp.call(void 0, );
  const file = pluginManager.getFile({ name: "operations", extName: ".ts", pluginKey });
  return /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _react.Parser, { language: "typescript", children: /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _react.File, { baseName: file.baseName, path: file.path, meta: file.meta, exportable: false, children: /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _react.File.Source, { children }) }) });
}
var defaultTemplates2 = { default: Template2, root: RootTemplate2 };
function Operations({ baseURL, Template: Template3 = defaultTemplates2.default }) {
  const operations = _hooks.useOperations.call(void 0, );
  return /* @__PURE__ */ _jsxruntime.jsx.call(void 0, Template3, { baseURL, name: "operations", operations });
}
Operations.File = function({ baseURL, ...props }) {
  const templates = { ...defaultTemplates2, ...props.templates };
  const Template3 = templates.default;
  const RootTemplate3 = templates.root;
  return /* @__PURE__ */ _jsxruntime.jsx.call(void 0, RootTemplate3, { children: /* @__PURE__ */ _jsxruntime.jsx.call(void 0, Operations, { baseURL, Template: Template3 }) });
};
Operations.templates = defaultTemplates2;




exports.Client = Client; exports.Operations = Operations;
