// @kubb/swagger-client/dist/index.js
import {
  Client,
  Operations
} from "./chunk-HOJIV7MX.js";

// src/plugin.ts
import path from "node:path";
import { FileManager, PluginManager, createPlugin } from "@kubb/core";
import { camelCase } from "@kubb/core/transformers";
import { renderTemplate } from "@kubb/core/utils";
import { pluginOasName } from "@kubb/plugin-oas";
import { getGroupedByTagFiles } from "@kubb/plugin-oas/utils";

// src/OperationGenerator.tsx
import { OperationGenerator as Generator } from "@kubb/plugin-oas";
import { Oas } from "@kubb/plugin-oas/components";
import { App, createRoot } from "@kubb/react";
import { jsx } from "@kubb/react/jsx-runtime";
var OperationGenerator = class extends Generator {
  async all(operations, _operationsByMethod) {
    const { pluginManager, oas, plugin, mode } = this.context;
    const root = createRoot({
      logger: pluginManager.logger
    });
    const templates = {
      operations: Operations.templates,
      client: Client.templates,
      ...this.options.templates
    };
    root.render(
      /* @__PURE__ */ jsx(App, { pluginManager, plugin, mode, children: /* @__PURE__ */ jsx(Oas, { oas, operations, generator: this, children: templates.operations && /* @__PURE__ */ jsx(Operations.File, { baseURL: this.options.baseURL, templates: templates.operations }) }) })
    );
    return root.files;
  }
  async operation(operation, options) {
    const { oas, pluginManager, plugin, mode } = this.context;
    const root = createRoot({
      logger: pluginManager.logger
    });
    const templates = {
      operations: Operations.templates,
      client: Client.templates,
      ...options.templates
    };
    if (!templates.client) {
      return [];
    }
    root.render(
      /* @__PURE__ */ jsx(App, { pluginManager, plugin: { ...plugin, options }, mode, children: /* @__PURE__ */ jsx(Oas, { oas, operations: [operation], generator: this, children: /* @__PURE__ */ jsx(Oas.Operation, { operation, children: /* @__PURE__ */ jsx(Client.File, { baseURL: this.options.baseURL, templates: templates.client }) }) }) })
    );
    return root.files;
  }
};

// src/plugin.ts
var pluginClientName = "plugin-client";
var pluginClient = createPlugin((options) => {
  const {
    output = { path: "clients" },
    group,
    exclude = [],
    include,
    override = [],
    transformers = {},
    dataReturnType = "data",
    pathParamsType = "inline",
    templates
  } = options;
  const template = group?.output ? group.output : `${output.path}/{{tag}}Controller`;
  return {
    name: pluginClientName,
    options: {
      extName: output.extName,
      dataReturnType,
      client: {
        importPath: "@kubb/swagger-client/client",
        ...options.client
      },
      pathParamsType,
      templates: {
        operations: Operations.templates,
        client: Client.templates,
        ...templates
      },
      baseURL: void 0
    },
    pre: [pluginOasName],
    resolvePath(baseName, pathMode, options2) {
      const root = path.resolve(this.config.root, this.config.output.path);
      const mode = pathMode ?? FileManager.getMode(path.resolve(root, output.path));
      if (mode === "single") {
        return path.resolve(root, output.path);
      }
      if (options2?.tag && group?.type === "tag") {
        const tag = camelCase(options2.tag);
        return path.resolve(root, renderTemplate(template, { tag }), baseName);
      }
      return path.resolve(root, output.path, baseName);
    },
    resolveName(name, type) {
      const resolvedName = camelCase(name, { isFile: type === "file" });
      if (type) {
        return transformers?.name?.(resolvedName, type) || resolvedName;
      }
      return resolvedName;
    },
    async writeFile(path2, source) {
      if (!source) {
        return;
      }
      return this.fileManager.write(path2, source, { sanity: false });
    },
    async buildStart() {
      const [swaggerPlugin] = PluginManager.getDependedPlugins(this.plugins, [pluginOasName]);
      const oas = await swaggerPlugin.api.getOas();
      const root = path.resolve(this.config.root, this.config.output.path);
      const mode = FileManager.getMode(path.resolve(root, output.path));
      const baseURL = await swaggerPlugin.api.getBaseURL();
      const operationGenerator = new OperationGenerator(
        {
          ...this.plugin.options,
          baseURL
        },
        {
          oas,
          pluginManager: this.pluginManager,
          plugin: this.plugin,
          contentType: swaggerPlugin.api.contentType,
          exclude,
          include,
          override,
          mode
        }
      );
      const files = await operationGenerator.build();
      await this.addFile(...files);
    },
    async buildEnd() {
      if (this.config.output.write === false) {
        return;
      }
      const root = path.resolve(this.config.root, this.config.output.path);
      if (group?.type === "tag") {
        const rootFiles = await getGroupedByTagFiles({
          logger: this.logger,
          files: this.fileManager.files,
          plugin: this.plugin,
          template,
          exportAs: group.exportAs || "{{tag}}Service",
          root,
          output
        });
        await this.addFile(...rootFiles);
      }
      await this.fileManager.addIndexes({
        root,
        output,
        meta: { pluginKey: this.plugin.key },
        logger: this.logger
      });
    }
  };
});

// src/index.ts
var definePluginDefault = pluginClient;
var definePlugin = pluginClient;
var src_default = definePluginDefault;
export {
  src_default as default,
  definePlugin,
  pluginClient,
  pluginClientName
};
