// backend-nest/src/modules/transports/transports.service.ts
import { Injectable, NotFoundException, BadRequestException, Logger, ConflictException } from '@nestjs/common';
import { Cron } from '@nestjs/schedule';
import { PrismaService } from '../../database/prisma.service';
import { TransportStatus, VehicleStatus, ProductStatus } from '@prisma/client';
import { CreateTransportDto } from './dto/create-transport.dto';
import { UpdateTransportDto } from './dto/update-transport.dto';
import { FilterTransportDto } from './dto/filter-transport.dto';
import { NotificationsService } from '../notifications/notifications.service';
import { AuditLogService } from '../audit-log/audit-log.service';

@Injectable()
export class TransportsService {
  private readonly logger = new Logger(TransportsService.name);

  constructor(
    private prisma: PrismaService,
    private notificationsService: NotificationsService,
    private auditLogService: AuditLogService,
  ) {}

  // ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
  // 🔒 MÉTODO AUXILIAR: Verificar se veículo está disponível
  // ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  private async checkVehicleAvailability(vehicleId: string, companyId: string) {
    this.logger.log(`🔍 Verificando disponibilidade do veículo ${vehicleId}...`);

    const vehicle = await this.prisma.vehicle.findFirst({
      where: { 
        id: vehicleId,
        companyId 
      }
    });

    if (!vehicle) {
      throw new NotFoundException('🚫 Veículo não encontrado ou não pertence a esta empresa');
    }

    if (vehicle.status === VehicleStatus.in_maintenance) {
      throw new BadRequestException(
        `🔧 Veículo ${vehicle.licensePlate} está em manutenção e não pode ser usado.\n` +
        `Por favor, escolha outro veículo.`
      );
    }

    if (vehicle.status === VehicleStatus.retired) {
      throw new BadRequestException(
        `❌ Veículo ${vehicle.licensePlate} está inativo.\n` +
        `Por favor, escolha outro veículo.`
      );
    }

    const activeTransport = await this.prisma.transport.findFirst({
      where: {
        vehicleId: vehicleId,
        status: {
          in: [TransportStatus.PENDING, TransportStatus.IN_TRANSIT, TransportStatus.ARRIVED],
        },
      },
      select: {
        id: true,
        internalCode: true,
        status: true,
        origin: true,
        destination: true,
        estimatedArrival: true,
      },
    });

    if (activeTransport) {
      this.logger.warn(`🚨 Veículo ${vehicle.licensePlate} já está em uso!`);
      this.logger.warn(`   Transporte ativo: ${activeTransport.internalCode}`);
      this.logger.warn(`   Status: ${activeTransport.status}`);
      this.logger.warn(`   Rota: ${activeTransport.origin} → ${activeTransport.destination}`);

      const statusMessages = {
        [TransportStatus.PENDING]: 'pendente de partida',
        [TransportStatus.IN_TRANSIT]: 'em trânsito',
        [TransportStatus.ARRIVED]: 'chegou ao destino e aguarda conferência',
      };

      throw new ConflictException(
        `🚫 VEÍCULO INDISPONÍVEL\n\n` +
        `🚗 Veículo: ${vehicle.licensePlate} (${vehicle.model})\n` +
        `📦 Já está alocado ao transporte: ${activeTransport.internalCode}\n` +
        `🚛 Status atual: ${statusMessages[activeTransport.status]}\n` +
        `📍 Rota: ${activeTransport.origin} → ${activeTransport.destination}\n` +
        `📅 Previsão de chegada: ${new Date(activeTransport.estimatedArrival).toLocaleDateString('pt-PT')}\n\n` +
        `💡 O veículo estará disponível após:\n` +
        `  • A entrega ser finalizada (status DELIVERED), ou\n` +
        `  • O transporte ser cancelado (status CANCELED)\n\n` +
        `🔍 Por favor, escolha outro veículo disponível.`
      );
    }

    this.logger.log(`✅ Veículo ${vehicle.licensePlate} está disponível!`);
    return vehicle;
  }

  // ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
  // 🤖 CRON JOB: Auto-Arrive Transports
  // ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  @Cron('0 5 0 * * *', {
    name: 'auto-arrive-transports',
    timeZone: 'Europe/Lisbon',
  })
  async autoArriveTransports() {
    this.logger.log('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━');
    this.logger.log('🤖 CRON JOB EXECUTADO: Auto-Arrive Transports');
    this.logger.log(`📅 Data/Hora: ${new Date().toLocaleString('pt-PT')}`);
    this.logger.log('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━');

    try {
      const today = new Date();
      today.setHours(0, 0, 0, 0);

      this.logger.log(`🔍 Procurando transportes IN_TRANSIT com data estimada <= ${today.toLocaleDateString('pt-PT')}`);

      const transportsToArrive = await this.prisma.transport.findMany({
        where: {
          status: TransportStatus.IN_TRANSIT,
          estimatedArrival: {
            lte: today,
          },
        },
        include: {
          vehicle: true,
          company: true,
          products: {
            include: {
              product: true,
            },
          },
        },
      });

      if (transportsToArrive.length === 0) {
        this.logger.log('✅ Nenhum transporte para processar hoje');
        this.logger.log('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━');
        return;
      }

      this.logger.log(`📦 Encontrados ${transportsToArrive.length} transporte(s) para processar`);

      let processedCount = 0;
      let errorCount = 0;

      for (const transport of transportsToArrive) {
        try {
          this.logger.log(`\n🚚 Processando transporte ${transport.internalCode}`);
          this.logger.log(`   📍 Rota: ${transport.origin} → ${transport.destination}`);
          this.logger.log(`   🚗 Veículo: ${transport.vehicle.licensePlate}`);
          this.logger.log(`   📅 Data estimada: ${new Date(transport.estimatedArrival).toLocaleDateString('pt-PT')}`);

          await this.prisma.transport.update({
            where: { id: transport.id },
            data: { status: TransportStatus.ARRIVED },
          });

          this.logger.log(`   ✅ Status mudado: IN_TRANSIT → ARRIVED`);

          const usersToNotify = await this.prisma.user.findMany({
            where: {
              companyId: transport.companyId,
              role: { in: ['ADMIN', 'OPERATOR'] },
              isActive: true,
            },
            select: {
              id: true,
              name: true,
              email: true,
            },
          });

          this.logger.log(`   📧 Enviando notificações para ${usersToNotify.length} utilizador(es)`);

          // CORREÇÃO AQUI: Adicionando o parâmetro origin faltante
          await this.notificationsService.notifyTransportArrived(
            transport.companyId,
            transport.internalCode,
            transport.origin, // Parâmetro adicionado
            transport.destination
          );

          this.logger.log(`      ✓ Notificações enviadas para ${usersToNotify.length} utilizadores`);

          processedCount++;
          this.logger.log(`   ✅ Transporte ${transport.internalCode} processado com sucesso`);

        } catch (error) {
          errorCount++;
          this.logger.error(`   ❌ Erro ao processar transporte ${transport.internalCode}: ${error.message}`);
        }
      }

      this.logger.log('\n━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━');
      this.logger.log('📊 RESUMO DO CRON JOB');
      this.logger.log(`   ✅ Processados: ${processedCount}`);
      this.logger.log(`   ❌ Erros: ${errorCount}`);
      this.logger.log(`   📦 Total: ${transportsToArrive.length}`);
      this.logger.log('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━');

    } catch (error) {
      this.logger.error('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━');
      this.logger.error('🚨 ERRO CRÍTICO NO CRON JOB');
      this.logger.error(`Mensagem: ${error.message}`);
      this.logger.error(`Stack: ${error.stack}`);
      this.logger.error('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━');
    }
  }

  // ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
  // 📦 CREATE - Criar novo transporte
  // ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  async create(data: CreateTransportDto, companyId: string, userId: string) {
    this.logger.log(`━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━`);
    this.logger.log(`📦 Criando novo transporte`);
    this.logger.log(`🏢 CompanyId: ${companyId}`);
    this.logger.log(`👤 UserId: ${userId}`);
    this.logger.log(`📦 Produtos: ${data.products?.length || 0}`);

    const vehicle = await this.checkVehicleAvailability(data.vehicleId, companyId);

    if (data.totalWeight > vehicle.capacity) {
      throw new BadRequestException(
        `⚖️ CAPACIDADE EXCEDIDA\n\n` +
        `🚗 Veículo: ${vehicle.licensePlate}\n` +
        `📊 Capacidade máxima: ${vehicle.capacity}kg\n` +
        `📦 Peso solicitado: ${data.totalWeight}kg\n` +
        `⚠️ Excesso: ${data.totalWeight - vehicle.capacity}kg\n\n` +
        `💡 Reduza a carga ou escolha um veículo maior.`
      );
    }

    if (data.products && data.products.length > 0) {
      for (const productData of data.products) {
        const product = await this.prisma.product.findFirst({
          where: {
            id: productData.productId,
            companyId,
          },
        });

        if (!product) {
          throw new NotFoundException(`Produto ${productData.productId} não encontrado`);
        }

        if (product.quantity < productData.quantity) {
          throw new BadRequestException(
            `📦 STOCK INSUFICIENTE\n\n` +
            `Produto: ${product.internalCode}\n` +
            `Disponível: ${product.quantity} un\n` +
            `Solicitado: ${productData.quantity} un\n` +
            `Faltam: ${productData.quantity - product.quantity} un`
          );
        }
      }
    }

    const lastTransport = await this.prisma.transport.findFirst({
      where: { companyId },
      orderBy: { createdAt: 'desc' }
    });

    const nextNumber = lastTransport 
      ? parseInt(lastTransport.internalCode.split('-')[1]) + 1 
      : 1;
    
    const internalCode = `TRP-${nextNumber.toString().padStart(6, '0')}`;
    this.logger.log(`🔢 Código gerado: ${internalCode}`);

    const transport = await this.prisma.$transaction(async (tx) => {
      const newTransport = await tx.transport.create({
        data: {
          internalCode,
          vehicleId: data.vehicleId,
          origin: data.origin,
          destination: data.destination,
          departureDate: new Date(data.departureDate),
          estimatedArrival: new Date(data.estimatedArrival),
          totalWeight: data.totalWeight,
          notes: data.notes,
          status: data.status || TransportStatus.PENDING,
          companyId,
        },
        include: {
          vehicle: true,
          company: true,
        },
      });

      this.logger.log(`✅ Transporte criado: ${newTransport.id}`);

      if (data.products && data.products.length > 0) {
        this.logger.log(`📦 Processando ${data.products.length} produto(s)...`);

        for (const productData of data.products) {
          await tx.transportProduct.create({
            data: {
              transportId: newTransport.id,
              productId: productData.productId,
              quantity: productData.quantity,
            },
          });

          this.logger.log(`  ✓ Produto ${productData.productId} associado`);

          const product = await tx.product.update({
            where: { id: productData.productId },
            data: {
              status: ProductStatus.DISPATCHED,
              quantity: {
                decrement: productData.quantity,
              },
            },
          });

          this.logger.log(`  ✓ Status mudado: ${product.status}`);
          this.logger.log(`  ✓ Quantidade atualizada: ${product.quantity}`);

          await tx.productMovement.create({
            data: {
              productId: productData.productId,
              previousStatus: ProductStatus.IN_STORAGE,
              newStatus: ProductStatus.DISPATCHED,
              quantity: productData.quantity,
              location: data.origin,
              reason: `Adicionado ao transporte ${internalCode}`,
              userId: userId,
            },
          });

          this.logger.log(`  ✓ Movimento registrado`);
        }
      }

      await tx.vehicle.update({
        where: { id: data.vehicleId },
        data: { status: VehicleStatus.in_use },
      });

      this.logger.log(`🚗 Veículo ${vehicle.licensePlate} → IN_USE (bloqueado)`);

      return newTransport;
    });

    this.logger.log(`━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━`);
    this.logger.log(`✅ Transporte criado com sucesso!`);
    this.logger.log(`   🆔 ID: ${transport.id}`);
    this.logger.log(`   🔢 Código: ${transport.internalCode}`);
    this.logger.log(`   📦 Produtos: ${data.products?.length || 0}`);
    this.logger.log(`   🚗 Veículo: ${transport.vehicle.licensePlate} → IN_USE`);
    this.logger.log(`━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━`);

    return transport;
  }

  async findAll(companyId?: string, filters?: FilterTransportDto) {
    const where: any = {};

    if (companyId) {
      where.companyId = companyId;
    }

    if (filters?.status) {
      where.status = filters.status;
    }

    if (filters?.startDate || filters?.endDate) {
      where.departureDate = {};
      if (filters.startDate) {
        where.departureDate.gte = new Date(filters.startDate);
      }
      if (filters.endDate) {
        where.departureDate.lte = new Date(filters.endDate);
      }
    }

    if (filters?.vehicleId) {
      where.vehicleId = filters.vehicleId;
    }

    return this.prisma.transport.findMany({
      where,
      include: {
        vehicle: true,
        company: true,
        products: {
          include: {
            product: {
              include: {
                supplier: true,
              },
            },
          },
        },
      },
      orderBy: { createdAt: 'desc' },
    });
  }

  async findPending(companyId?: string) {
    const where: any = {
      status: TransportStatus.PENDING,
    };

    if (companyId) {
      where.companyId = companyId;
    }

    return this.prisma.transport.findMany({
      where,
      include: {
        vehicle: true,
        company: true,
        products: {
          include: {
            product: true,
          },
        },
      },
      orderBy: { departureDate: 'asc' },
    });
  }

  async findInTransit(companyId?: string) {
    const where: any = {
      status: TransportStatus.IN_TRANSIT,
    };

    if (companyId) {
      where.companyId = companyId;
    }

    return this.prisma.transport.findMany({
      where,
      include: {
        vehicle: true,
        company: true,
        products: {
          include: {
            product: true,
          },
        },
      },
      orderBy: { departureDate: 'asc' },
    });
  }

  async findOne(id: string, companyId?: string) {
    const where: any = { id };
    
    if (companyId) {
      where.companyId = companyId;
    }

    const transport = await this.prisma.transport.findFirst({
      where,
      include: {
        vehicle: true,
        company: true,
        products: {
          include: {
            product: {
              include: {
                supplier: true,
              },
            },
          },
        },
      },
    });

    if (!transport) {
      throw new NotFoundException('Transporte não encontrado');
    }

    return transport;
  }

  async update(id: string, data: UpdateTransportDto, companyId?: string, userId?: string) {
    this.logger.log(`━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━`);
    this.logger.log(`🔄 Atualizando transporte ${id}`);
    this.logger.log(`👤 UserId: ${userId || 'system'}`);

    const transport = await this.findOne(id, companyId);

    const updateData: any = { ...data };
    
    if (data.departureDate) {
      updateData.departureDate = new Date(data.departureDate);
    }
    if (data.estimatedArrival) {
      updateData.estimatedArrival = new Date(data.estimatedArrival);
    }
    if (data.actualArrival) {
      updateData.actualArrival = new Date(data.actualArrival);
    }

    if (data.status === TransportStatus.DELIVERED && transport.status !== TransportStatus.DELIVERED) {
      this.logger.log(`✅ Status mudando para DELIVERED - Iniciando automação...`);

      if (transport.status !== TransportStatus.ARRIVED) {
        throw new BadRequestException(
          `⚠️ VALIDAÇÃO DE STATUS\n\n` +
          `Para marcar como DELIVERED, o transporte deve estar no status ARRIVED.\n` +
          `Status atual: ${transport.status}\n\n` +
          `💡 Fluxo correto: PENDING → IN_TRANSIT → ARRIVED → DELIVERED`
        );
      }

      if (!updateData.actualArrival) {
        updateData.actualArrival = new Date();
        this.logger.log(`📅 Data real de chegada definida automaticamente: ${updateData.actualArrival.toISOString()}`);
      }

      return await this.prisma.$transaction(async (tx) => {
        const updatedTransport = await tx.transport.update({
          where: { id },
          data: updateData,
          include: {
            vehicle: true,
            company: true,
            products: {
              include: {
                product: true,
              },
            },
          },
        });

        this.logger.log(`📦 Processando ${updatedTransport.products.length} produto(s)...`);

        for (const tp of updatedTransport.products) {
          await tx.product.update({
            where: { id: tp.productId },
            data: { 
              status: ProductStatus.APPROVED,
              currentLocation: transport.destination,
            },
          });

          this.logger.log(`  ✓ Produto ${tp.product.internalCode} → APPROVED`);

          await tx.productMovement.create({
            data: {
              productId: tp.productId,
              previousStatus: ProductStatus.DISPATCHED,
              newStatus: ProductStatus.APPROVED,
              quantity: tp.quantity,
              location: transport.destination,
              reason: `Transporte ${transport.internalCode} finalizado e conferido` +
                     (data.receivedBy ? ` por ${data.receivedBy}` : ''),
              userId: userId || 'system',
            },
          });
        }

        await tx.vehicle.update({
          where: { id: transport.vehicleId },
          data: { status: VehicleStatus.available },
        });

        this.logger.log(`🚗 Veículo ${updatedTransport.vehicle.licensePlate} → AVAILABLE (liberado)`);
        
        if (data.receivedBy) {
          this.logger.log(`👤 Recebido por: ${data.receivedBy}`);
        }
        if (data.receivingNotes) {
          this.logger.log(`📝 Observações: ${data.receivingNotes}`);
        }

        try {
          await this.notificationsService.notifyTransportDelivered(
            transport.companyId,
            transport.internalCode,
            data.receivedBy || 'Não especificado'
          );
        } catch (error) {
          this.logger.error(`❌ Erro ao enviar notificação de entrega: ${error.message}`);
        }
        
        this.logger.log(`━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━`);

        return updatedTransport;
      });
    }

    if (data.status === TransportStatus.CANCELED && transport.status !== TransportStatus.CANCELED) {
      this.logger.log(`❌ Status mudando para CANCELED - Revertendo automação...`);

      return await this.prisma.$transaction(async (tx) => {
        const updatedTransport = await tx.transport.update({
          where: { id },
          data: updateData,
          include: {
            vehicle: true,
            company: true,
            products: {
              include: {
                product: true,
              },
            },
          },
        });

        for (const tp of updatedTransport.products) {
          await tx.product.update({
            where: { id: tp.productId },
            data: { 
              status: ProductStatus.IN_STORAGE,
              quantity: {
                increment: tp.quantity,
              },
            },
          });

          this.logger.log(`  ✓ Produto ${tp.product.internalCode} → IN_STORAGE (devolvido ao stock)`);

          await tx.productMovement.create({
            data: {
              productId: tp.productId,
              previousStatus: ProductStatus.DISPATCHED,
              newStatus: ProductStatus.IN_STORAGE,
              quantity: tp.quantity,
              location: transport.origin,
              reason: `Transporte ${transport.internalCode} cancelado - produtos devolvidos`,
              userId: userId || 'system',
            },
          });
        }

        await tx.vehicle.update({
          where: { id: transport.vehicleId },
          data: { status: VehicleStatus.available },
        });

        this.logger.log(`🚗 Veículo ${updatedTransport.vehicle.licensePlate} → AVAILABLE (liberado)`);
        
        // CORREÇÃO AQUI: Usando userId fornecido ou 'system' como fallback
        const notificationUserId = userId || 'system';
        try {
          await this.notificationsService.create({
            title: '❌ Transporte Cancelado',
            message: `O transporte ${transport.internalCode} foi cancelado. ${data.notes || 'Sem motivo especificado'}`,
            companyId: transport.companyId,
            userId: notificationUserId, // Parâmetro adicionado
          });
        } catch (error) {
          this.logger.error(`❌ Erro ao enviar notificação de cancelamento: ${error.message}`);
        }
        
        this.logger.log(`━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━`);

        return updatedTransport;
      });
    }

    this.logger.log(`📝 Atualização simples de campos logísticos`);
    
    const result = await this.prisma.transport.update({
      where: { id },
      data: updateData,
      include: {
        vehicle: true,
        company: true,
        products: {
          include: {
            product: {
              include: {
                supplier: true,
              },
            },
          },
        },
      },
    });

    this.logger.log(`✅ Transporte atualizado com sucesso`);
    this.logger.log(`━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━`);

    return result;
  }

  async updateStatus(id: string, status: TransportStatus, companyId?: string, userId?: string) {
    if (!Object.values(TransportStatus).includes(status)) {
      throw new BadRequestException(
        `Status inválido. Valores aceites: ${Object.values(TransportStatus).join(', ')}`
      );
    }

    return this.update(id, { status }, companyId, userId);
  }

  async remove(id: string, companyId?: string, userId?: string, force = false) {
    this.logger.log(`━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━`);
    this.logger.log(`🗑️ Tentando remover transporte ${id} (force=${force})`);
    this.logger.log(`🏢 CompanyId: ${companyId || 'TODAS (SUPER_ADMIN)'}`);
    this.logger.log(`👤 UserId: ${userId || 'system'}`);

    try {
      const transport = await this.findOne(id, companyId);
      this.logger.log(`✅ Transporte encontrado: ${transport.internalCode}`);
      this.logger.log(`📊 Status atual: ${transport.status}`);
      this.logger.log(`📦 Produtos associados: ${transport.products.length}`);

      if (!force && (transport.status === TransportStatus.IN_TRANSIT || transport.status === TransportStatus.ARRIVED)) {
        this.logger.warn(`🚨 BLOQUEADO - Transporte em ${transport.status}`);
        
        throw new ConflictException(
          `🚫 Não é possível eliminar um transporte em trânsito ou que já chegou.\n\n` +
          `📦 Transporte: ${transport.internalCode}\n` +
          `🚛 Status: ${transport.status}\n` +
          `📍 Rota: ${transport.origin} → ${transport.destination}\n\n` +
          `💡 Pode eliminar o transporte quando:\n` +
          `  • Finalizar a entrega (status DELIVERED)\n` +
          `  • Cancelar o transporte (status CANCELED)`
        );
      }

      if (!force && transport.status === TransportStatus.DELIVERED) {
        this.logger.warn(`🚨 BLOQUEADO - Transporte já entregue (dados históricos)`);
        
        throw new ConflictException(
          `🚫 Não é possível eliminar um transporte já entregue.\n\n` +
          `📦 Transporte: ${transport.internalCode}\n` +
          `🚛 Status: DELIVERED\n` +
          `📍 Rota: ${transport.origin} → ${transport.destination}\n` +
          `📅 Data de entrega: ${new Date(transport.estimatedArrival).toLocaleDateString('pt-PT')}\n\n` +
          `ℹ️ Transportes entregues são mantidos para histórico e auditoria.\n` +
          `Se necessário, pode cancelar o transporte antes de eliminá-lo.`
        );
      }

      if (transport.status === TransportStatus.PENDING && transport.products.length > 0) {
        this.logger.log(`⚠️ Transporte PENDING com ${transport.products.length} produto(s)`);
        this.logger.log(`🔄 Devolvendo produtos ao stock...`);

        await this.prisma.$transaction(async (tx) => {
          const productsList = transport.products
            .map(tp => `${tp.product.internalCode} (${tp.quantity} un)`)
            .join(', ');

          this.logger.log(`📦 Produtos a devolver: ${productsList}`);

          for (const tp of transport.products) {
            await tx.product.update({
              where: { id: tp.productId },
              data: {
                status: ProductStatus.IN_STORAGE,
                quantity: {
                  increment: tp.quantity,
                },
              },
            });

            this.logger.log(`  ✓ ${tp.product.internalCode}: +${tp.quantity} un → IN_STORAGE`);

            await tx.productMovement.create({
              data: {
                productId: tp.productId,
                previousStatus: ProductStatus.DISPATCHED,
                newStatus: ProductStatus.IN_STORAGE,
                quantity: tp.quantity,
                location: transport.origin,
                reason: `Transporte ${transport.internalCode} eliminado - produtos devolvidos ao stock`,
                userId: userId || 'system',
              },
            });
          }

          await tx.vehicle.update({
            where: { id: transport.vehicleId },
            data: { status: VehicleStatus.available },
          });

          this.logger.log(`🚗 Veículo ${transport.vehicle.licensePlate} → AVAILABLE (liberado)`);

          await tx.transportProduct.deleteMany({
            where: { transportId: id },
          });

          this.logger.log(`🗑️ Relações produto-transporte eliminadas`);

          await tx.transport.delete({
            where: { id },
          });

          this.logger.log(`✅ Transporte eliminado com sucesso`);
        });

        this.logger.log(`━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━`);
        this.logger.log(`✅ Transporte "${transport.internalCode}" eliminado`);
        this.logger.log(`   📦 ${transport.products.length} produto(s) devolvido(s) ao stock`);
        this.logger.log(`   🚗 Veículo liberado`);
        this.logger.log(`━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━`);

        return { 
          message: `✅ Transporte "${transport.internalCode}" eliminado com sucesso`,
          productsReturned: transport.products.length,
          details: `${transport.products.length} produto(s) devolvido(s) ao stock`
        };
      }

      if (transport.status === TransportStatus.CANCELED || 
         (transport.status === TransportStatus.PENDING && transport.products.length === 0)) {
        
        this.logger.log(`✅ Transporte pode ser eliminado (${transport.status}, sem produtos ativos)`);

        await this.prisma.$transaction(async (tx) => {
          if (transport.status === TransportStatus.PENDING) {
            await tx.vehicle.update({
              where: { id: transport.vehicleId },
              data: { status: VehicleStatus.available },
            });
            this.logger.log(`🚗 Veículo ${transport.vehicle.licensePlate} → AVAILABLE (liberado)`);
          }

          await tx.transportProduct.deleteMany({
            where: { transportId: id },
          });

          await tx.transport.delete({
            where: { id },
          });
        });

        this.logger.log(`━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━`);
        this.logger.log(`✅ Transporte "${transport.internalCode}" eliminado com sucesso`);
        this.logger.log(`━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━`);

        return { 
          message: `✅ Transporte "${transport.internalCode}" eliminado com sucesso`
        };
      }

      throw new BadRequestException(
        `Estado inesperado do transporte. Status: ${transport.status}`
      );

    } catch (error) {
      if (error instanceof NotFoundException || 
          error instanceof ConflictException || 
          error instanceof BadRequestException) {
        throw error;
      }

      this.logger.error(`❌ Erro inesperado ao eliminar transporte ${id}`);
      this.logger.error(`Mensagem: ${error.message}`);
      this.logger.error(`Stack: ${error.stack}`);
      
      throw new BadRequestException(
        `Erro ao eliminar transporte: ${error.message}`
      );
    }
  }

  async getStats(companyId?: string) {
    const where: any = {};
    
    if (companyId) {
      where.companyId = companyId;
    }

    const [total, pending, inTransit, arrived, delivered, cancelled] = await Promise.all([
      this.prisma.transport.count({ where }),
      this.prisma.transport.count({ where: { ...where, status: TransportStatus.PENDING } }),
      this.prisma.transport.count({ where: { ...where, status: TransportStatus.IN_TRANSIT } }),
      this.prisma.transport.count({ where: { ...where, status: TransportStatus.ARRIVED } }),
      this.prisma.transport.count({ where: { ...where, status: TransportStatus.DELIVERED } }),
      this.prisma.transport.count({ where: { ...where, status: TransportStatus.CANCELED } }),
    ]);

    return {
      total,
      pending,
      inTransit,
      arrived,
      delivered,
      cancelled,
    };
  }

  // ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
  // 🗺️ MAPEAMENTO DE CIDADES PARA COORDENADAS GPS
  // ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
  
  private readonly citiesCoordinates: Record<string, { lat: number; lng: number }> = {
    'lisboa': { lat: 38.7223, lng: -9.1393 },
    'lisbon': { lat: 38.7223, lng: -9.1393 },
    'porto': { lat: 41.1579, lng: -8.6291 },
    'oporto': { lat: 41.1579, lng: -8.6291 },
    'braga': { lat: 41.5455, lng: -8.4268 },
    'covilhã': { lat: 40.2848, lng: -7.5025 },
    'covilha': { lat: 40.2848, lng: -7.5025 },
    'aveiro': { lat: 40.6384, lng: -8.6488 },
    'évora': { lat: 38.6644, lng: -8.3026 },
    'evora': { lat: 38.6644, lng: -8.3026 },
    'leiria': { lat: 39.7471, lng: -8.8067 },
    'santarém': { lat: 39.2227, lng: -8.6898 },
    'santarem': { lat: 39.2227, lng: -8.6898 },
    'castelo branco': { lat: 40.2863, lng: -7.5006 },
    'guarda': { lat: 40.5365, lng: -7.2714 },
    'belmonte': { lat: 40.3542, lng: -7.3889 },
    'viseu': { lat: 40.6642, lng: -7.2686 },
    'vila real': { lat: 41.2925, lng: -7.7433 },
    'vila-real': { lat: 41.2925, lng: -7.7433 },
    'bragança': { lat: 41.8047, lng: -6.7591 },
    'braganca': { lat: 41.8047, lng: -6.7591 },
    'funchal': { lat: 32.6542, lng: -16.9045 },
    'ponta delgada': { lat: 37.7412, lng: -25.6756 },
    'amadora': { lat: 38.7620, lng: -9.2360 },
    'barreiro': { lat: 38.6636, lng: -9.0759 },
    'carcavelos': { lat: 38.6542, lng: -9.3167 },
    'cascais': { lat: 38.6954, lng: -9.4202 },
    'caparica': { lat: 38.6576, lng: -9.2141 },
    'loures': { lat: 38.8268, lng: -9.1583 },
    'oeiras': { lat: 38.6821, lng: -9.3102 },
    'setúbal': { lat: 38.5246, lng: -8.8885 },
    'setubal': { lat: 38.5246, lng: -8.8885 },
    'almada': { lat: 38.6813, lng: -9.2138 },
    'cacem': { lat: 38.7537, lng: -9.3051 },
    'cacém': { lat: 38.7537, lng: -9.3051 },
    'alcântara': { lat: 38.7094, lng: -9.1838 },
    'alcantara': { lat: 38.7094, lng: -9.1838 },
    'belém': { lat: 38.7029, lng: -9.2127 },
    'belem': { lat: 38.7029, lng: -9.2127 },
    'azambuja': { lat: 39.0766, lng: -8.8428 },
    'alcochete': { lat: 38.7589, lng: -8.9787 },
  };

  private getCityCoordinates(cityName: string): { lat: number; lng: number } {
    const normalizedName = cityName.trim().toLowerCase().normalize('NFD').replace(/[\u0300-\u036f]/g, '');
    const coords = this.citiesCoordinates[normalizedName];
    
    if (!coords) {
      // Se a cidade não for encontrada, usar Lisboa como fallback
      this.logger.warn(`⚠️ Cidade '${cityName}' não encontrada. Usando Lisboa como fallback.`);
      return { lat: 38.7223, lng: -9.1393 };
    }
    
    return coords;
  }

  // ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
  // 🗺️ RASTREAMENTO GPS - NOVOS MÉTODOS
  // ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  async getTrackingRoutes(companyId?: string) {
    this.logger.log(`📍 Obtendo rotas de rastreamento GPS`);
    
    const where: any = {};
    
    if (companyId) {
      where.companyId = companyId;
    }

    const transports = await this.prisma.transport.findMany({
      where,
      include: {
        vehicle: true,
        company: true,
      },
      orderBy: { departureDate: 'desc' },
    });

    // Mapear transportes para rotas de rastreamento
    const trackingRoutes = transports.map(t => {
      const originCoords = this.getCityCoordinates(t.origin);
      const destCoords = this.getCityCoordinates(t.destination);
      const midpointLat = (originCoords.lat + destCoords.lat) / 2;
      const midpointLng = (originCoords.lng + destCoords.lng) / 2;
      
      // Garantir que as coordenadas são números válidos
      const safeOriginLat = Number(originCoords.lat) || 39.5;
      const safeOriginLng = Number(originCoords.lng) || -8.0;
      const safeDestLat = Number(destCoords.lat) || 39.5;
      const safeDestLng = Number(destCoords.lng) || -8.0;
      const safeMidLat = (safeOriginLat + safeDestLat) / 2;
      const safeMidLng = (safeOriginLng + safeDestLng) / 2;

      this.logger.debug(`📍 Rota: ${t.origin} → ${t.destination}`);
      this.logger.debug(`   Origem: [${safeOriginLat}, ${safeOriginLng}]`);
      this.logger.debug(`   Destino: [${safeDestLat}, ${safeDestLng}]`);
      
      return {
        id: t.id,
        name: `${t.origin} → ${t.destination}`,
        origin: t.origin,
        destination: t.destination,
        // Incluir coordenadas individuais também (backward compatibility)
        origin_lat: safeOriginLat,
        origin_lng: safeOriginLng,
        destination_lat: safeDestLat,
        destination_lng: safeDestLng,
        status: t.status.toLowerCase(),
        startTime: t.departureDate,
        endTime: t.estimatedArrival,
        actualArrival: t.actualArrival,
        // Locations é o array esperado pelo TrackingMap
        locations: [
          {
            lat: safeOriginLat,
            lng: safeOriginLng,
            timestamp: t.departureDate.toISOString(),
            speed: 0,
          },
          {
            lat: safeMidLat,
            lng: safeMidLng,
            timestamp: new Date(
              t.departureDate.getTime() + 
              (t.estimatedArrival.getTime() - t.departureDate.getTime()) / 2
            ).toISOString(),
            speed: 50,
          },
          {
            lat: safeDestLat,
            lng: safeDestLng,
            timestamp: (t.actualArrival || t.estimatedArrival).toISOString(),
            speed: 0,
          },
        ],
        vehicle: t.vehicle,
        vehicleId: t.vehicleId,
        transportId: t.id, // Adicionar também transportId para compatibilidade
        transport: t.vehicle ? { id: t.vehicleId } : undefined,
        company: t.company,
      };
    });

    this.logger.log(`✅ ${trackingRoutes.length} rota(s) encontrada(s) com coordenadas GPS`);
    
    return trackingRoutes;
  }

  async deleteTrackingRoute(id: string, companyId?: string, userId?: string) {
    this.logger.log(`🗑️ Eliminando rota de rastreamento ${id}`);
    
    const transport = await this.findOne(id, companyId);
    
    this.logger.log(`✅ Transporte encontrado: ${transport.internalCode}`);
    
    // Aqui poderíamos eliminar dados de GPS do banco se existissem
    // Por enquanto, apenas retornamos sucesso
    
    this.logger.log(`✅ Rota de rastreamento ${id} eliminada`);
    
    return {
      message: `✅ Rota de rastreamento do transporte ${transport.internalCode} eliminada`,
      transportId: id,
    };
  }

  async clearAllTrackingRoutes(companyId?: string, userId?: string) {
    this.logger.log(`🗑️ Limpando todo o rastreamento GPS`);
    this.logger.log(`🏢 CompanyId: ${companyId || 'TODAS (SUPER_ADMIN)'}`);
    this.logger.log(`👤 UserId: ${userId || 'system'}`);
    
    const where: any = {};
    
    if (companyId) {
      where.companyId = companyId;
    }

    const count = await this.prisma.transport.count({ where });
    
    this.logger.log(`📊 Transportes afetados: ${count}`);
    
    // Aqui poderíamos limpar dados de GPS do banco se existissem
    // Por enquanto, apenas registramos a ação e retornamos sucesso
    
    this.logger.log(`✅ Todo o rastreamento GPS foi limpo`);
    
    return {
      message: `✅ Todo o rastreamento GPS foi eliminado com sucesso`,
      routesDeleted: count,
    };
  }
}
