// @kubb/cli/dist/chunk-M2AU2EGM.js
// src/utils/spinner.ts
import ora from "ora";
var spinner = ora({
  spinner: "clock"
});

// src/generate.ts
import { LogLevel as LogLevel2, LogMapper, createLogger, randomCliColour as randomCliColour2 } from "@kubb/core/logger";
import c4 from "tinyrainbow";
import { Warning, safeBuild } from "@kubb/core";
import { createConsola } from "consola";

// src/utils/executeHooks.ts
import { LogLevel } from "@kubb/core/logger";
import { execa } from "execa";
import { parseArgsStringToArgv } from "string-argv";
import c2 from "tinyrainbow";

// src/utils/OraWritable.ts
import { Writable } from "node:stream";
import c from "tinyrainbow";
var OraWritable = class extends Writable {
  constructor(spinner2, command, opts) {
    super(opts);
    this.command = command;
    this.spinner = spinner2;
  }
  _write(chunk, _encoding, callback) {
    this.spinner.suffixText = `

${c.bold(c.blue(this.command))}: ${chunk?.toString()}`;
    callback();
  }
};

// src/utils/executeHooks.ts
async function executeHooks({ hooks, logLevel }) {
  if (!hooks?.done) {
    return;
  }
  const commands = Array.isArray(hooks.done) ? hooks.done : [hooks.done];
  if (logLevel === LogLevel.silent) {
    spinner.start("Executing hooks");
  }
  const executers = commands.map(async (command) => {
    const oraWritable = new OraWritable(spinner, command);
    const abortController = new AbortController();
    const [cmd, ..._args] = [...parseArgsStringToArgv(command)];
    if (!cmd) {
      return null;
    }
    spinner.start(`Executing hook ${logLevel !== "silent" ? c2.dim(command) : ""}`);
    const subProcess = await execa(cmd, _args, {
      detached: true,
      cancelSignal: abortController.signal,
      stdout: ["pipe", oraWritable]
    });
    spinner.suffixText = "";
    if (logLevel === LogLevel.silent) {
      spinner.succeed(`Executing hook ${logLevel !== "silent" ? c2.dim(command) : ""}`);
      if (subProcess) {
        console.log(subProcess.stdout);
      }
    }
    oraWritable.destroy();
    return { subProcess, abort: abortController.abort.bind(abortController) };
  }).filter(Boolean);
  await Promise.all(executers);
  if (logLevel === LogLevel.silent) {
    spinner.succeed("Executing hooks");
  }
}

// src/utils/getErrorCauses.ts
function getErrorCauses(errors) {
  return errors.reduce((prev, error) => {
    const causedError = error?.cause;
    if (causedError) {
      prev = [...prev, ...getErrorCauses([causedError])];
      return prev;
    }
    prev = [...prev, error];
    return prev;
  }, []).filter(Boolean);
}

// src/utils/getSummary.ts
import path from "node:path";
import { randomCliColour } from "@kubb/core/logger";
import c3 from "tinyrainbow";

// src/utils/parseHrtimeToSeconds.ts
function parseHrtimeToSeconds(hrtime) {
  const seconds = (hrtime[0] + hrtime[1] / 1e9).toFixed(3);
  return seconds;
}

// src/utils/getSummary.ts
function getSummary({ pluginManager, status, hrstart, config, logger }) {
  const { logLevel } = logger;
  const logs = [];
  const elapsedSeconds = parseHrtimeToSeconds(process.hrtime(hrstart));
  const buildStartPlugins = pluginManager.executed.filter((item) => item.hookName === "buildStart" && item.plugin.name !== "core").map((item) => item.plugin.name);
  const buildEndPlugins = pluginManager.executed.filter((item) => item.hookName === "buildEnd" && item.plugin.name !== "core").map((item) => item.plugin.name);
  const failedPlugins = config.plugins?.filter((plugin) => !buildEndPlugins.includes(plugin.name))?.map((plugin) => plugin.name);
  const pluginsCount = config.plugins?.length || 0;
  const files = pluginManager.fileManager.files.sort((a, b) => {
    if (!a.meta?.pluginKey?.[0] || !b.meta?.pluginKey?.[0]) {
      return 0;
    }
    if (a.meta?.pluginKey?.[0]?.length < b.meta?.pluginKey?.[0]?.length) {
      return 1;
    }
    if (a.meta?.pluginKey?.[0]?.length > b.meta?.pluginKey?.[0]?.length) {
      return -1;
    }
    return 0;
  });
  const meta = {
    name: config.name,
    plugins: status === "success" ? `${c3.green(`${buildStartPlugins.length} successful`)}, ${pluginsCount} total` : `${c3.red(`${failedPlugins?.length ?? 1} failed`)}, ${pluginsCount} total`,
    pluginsFailed: status === "failed" ? failedPlugins?.map((name) => randomCliColour(name))?.join(", ") : void 0,
    filesCreated: files.length,
    time: c3.yellow(`${elapsedSeconds}s`),
    endTime: c3.yellow(Date()),
    output: path.isAbsolute(config.root) ? path.resolve(config.root, config.output.path) : config.root
  };
  logger.emit("debug", ["\nGenerated files:\n"]);
  logger.emit(
    "debug",
    files.map((file) => `${randomCliColour(JSON.stringify(file.meta?.pluginKey))} ${file.path}`)
  );
  logs.push(
    [
      ["\n", true],
      [`     ${c3.bold("Name:")}      ${meta.name}`, !!meta.name],
      [`  ${c3.bold("Plugins:")}      ${meta.plugins}`, true],
      [`   ${c3.dim("Failed:")}      ${meta.pluginsFailed || "none"}`, !!meta.pluginsFailed],
      [`${c3.bold("Generated:")}      ${meta.filesCreated} files`, true],
      [`     ${c3.bold("Time:")}      ${meta.time}`, true],
      [`    ${c3.bold("Ended:")}      ${meta.endTime}`, true],
      [`   ${c3.bold("Output:")}      ${meta.output}`, true]
    ].map((item) => {
      if (item.at(1)) {
        return item.at(0);
      }
      return void 0;
    }).filter(Boolean).join("\n")
  );
  return logs;
}

// src/utils/writeLog.ts
import { resolve } from "node:path";
import { read, write } from "@kubb/fs";
async function writeLog(data) {
  if (data.trim() === "") {
    return void 0;
  }
  const path2 = resolve(process.cwd(), "kubb-log.log");
  let previousLogs = "";
  try {
    previousLogs = await read(resolve(path2));
  } catch (_err) {
  }
  return write(path2, [previousLogs, data.trim()].filter(Boolean).join("\n\n\n"), { sanity: false });
}

// src/generate.ts
async function generate({ input, config, args }) {
  const logLevel = args.logLevel || LogLevel2.silent;
  const logger = createLogger({
    logLevel,
    name: config.name,
    spinner,
    consola: createConsola({
      level: LogMapper[logLevel] || 3
    })
  });
  logger.on("debug", async (messages) => {
    if (logLevel === LogLevel2.debug) {
      await writeLog(messages.join("\n"));
    }
  });
  logger.consola?.wrapConsole();
  if (logger.name) {
    spinner.prefixText = randomCliColour2(logger.name);
  }
  const hrstart = process.hrtime();
  if (args.logLevel === LogLevel2.debug) {
    const { performance, PerformanceObserver } = await import("node:perf_hooks");
    const performanceOpserver = new PerformanceObserver((items) => {
      const message = `${items.getEntries()[0]?.duration.toFixed(0)}ms`;
      spinner.suffixText = c4.yellow(message);
      performance.clearMarks();
    });
    performanceOpserver.observe({ type: "measure" });
  }
  const { root = process.cwd(), ...userConfig } = config;
  const inputPath = input ?? ("path" in userConfig.input ? userConfig.input.path : void 0);
  spinner.start(`\u{1F680} Building ${logLevel !== "silent" ? c4.dim(inputPath) : ""}`);
  const definedConfig = {
    root,
    ...userConfig,
    input: inputPath ? {
      ...userConfig.input,
      path: inputPath
    } : userConfig.input,
    output: {
      write: true,
      ...userConfig.output
    }
  };
  const { pluginManager, error } = await safeBuild({
    config: definedConfig,
    logger
  });
  const summary = getSummary({
    pluginManager,
    config: definedConfig,
    status: error ? "failed" : "success",
    hrstart,
    logger
  });
  if (error) {
    spinner.suffixText = "";
    spinner.fail(`\u{1F680} Build failed ${logLevel !== "silent" ? c4.dim(inputPath) : ""}`);
    console.log(summary.join(""));
    if (error instanceof Warning) {
      spinner.warn(c4.yellow(error.message));
      process.exit(0);
    }
    const errors = getErrorCauses([error]);
    if (logger.consola && errors.length && logLevel === LogLevel2.debug) {
      errors.forEach((err) => {
        logger.consola.error(err);
      });
    }
    logger.consola?.error(error);
    process.exit(0);
  }
  await executeHooks({ hooks: config.hooks, logLevel });
  spinner.suffixText = "";
  spinner.succeed(`\u{1F680} Build completed ${logLevel !== "silent" ? c4.dim(inputPath) : ""}`);
  console.log(summary.join(""));
}

export {
  spinner,
  generate
};
