// @kubb/cli/dist/generate-WRHF6H6X.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { newObj[key] = obj[key]; } } } newObj.default = obj; return newObj; } } function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; } function _optionalChain(ops) { let lastAccessLHS = undefined; let value = ops[0]; let i = 1; while (i < ops.length) { const op = ops[i]; const fn = ops[i + 1]; i += 2; if ((op === 'optionalAccess' || op === 'optionalCall') && value == null) { return undefined; } if (op === 'access' || op === 'optionalAccess') { lastAccessLHS = value; value = fn(value); } else if (op === 'call' || op === 'optionalCall') { value = fn((...args) => value.call(lastAccessLHS, ...args)); lastAccessLHS = undefined; } } return value; }





var _chunkSYMMPQK7cjs = require('./chunk-SYMMPQK7.cjs');

// src/commands/generate.ts
_chunkSYMMPQK7cjs.init_cjs_shims.call(void 0, );
var _logger = require('@kubb/core/logger');
var _citty = require('citty');
var _path = require('path'); var _path2 = _interopRequireDefault(_path);

// src/utils/getConfig.ts
_chunkSYMMPQK7cjs.init_cjs_shims.call(void 0, );
var _utils = require('@kubb/core/utils');

// src/utils/getPlugins.ts
_chunkSYMMPQK7cjs.init_cjs_shims.call(void 0, );
var _core = require('@kubb/core');
function isJSONPlugins(plugins) {
  return !!_optionalChain([plugins, 'optionalAccess', _2 => _2.some, 'call', _3 => _3((plugin) => {
    return Array.isArray(plugin) && typeof _optionalChain([plugin, 'optionalAccess', _4 => _4.at, 'call', _5 => _5(0)]) === "string";
  })]);
}
function isObjectPlugins(plugins) {
  return plugins instanceof Object && !Array.isArray(plugins);
}
async function importPlugin(name, options) {
  const packageManager = new (0, _core.PackageManager)(process.cwd());
  const importedPlugin = process.env.NODE_ENV === "test" ? await Promise.resolve().then(() => _interopRequireWildcard(require(name))) : await packageManager.import(name);
  return _optionalChain([importedPlugin, 'optionalAccess', _6 => _6.default]) ? importedPlugin.default(options) : importedPlugin(options);
}
function getPlugins(plugins) {
  if (isObjectPlugins(plugins)) {
    throw new Error("Object plugins are not supported anymore, best to use http://kubb.dev/guide/configure#json");
  }
  if (isJSONPlugins(plugins)) {
    const jsonPlugins = plugins;
    const promises = jsonPlugins.map((plugin) => {
      const [name, options = {}] = plugin;
      return importPlugin(name, options);
    });
    return Promise.all(promises);
  }
  return Promise.resolve(plugins);
}

// src/utils/getConfig.ts
async function getConfig(result, args2) {
  const config = _optionalChain([result, 'optionalAccess', _7 => _7.config]);
  let kubbUserConfig = Promise.resolve(config);
  if (typeof config === "function") {
    const possiblePromise = config(args2);
    if (_utils.isPromise.call(void 0, possiblePromise)) {
      kubbUserConfig = possiblePromise;
    }
    kubbUserConfig = Promise.resolve(possiblePromise);
  }
  let JSONConfig = await kubbUserConfig;
  if (Array.isArray(JSONConfig)) {
    const promises = JSONConfig.map(async (item) => {
      return {
        ...item,
        plugins: item.plugins ? await getPlugins(item.plugins) : void 0
      };
    });
    return Promise.all(promises);
  }
  JSONConfig = {
    ...JSONConfig,
    plugins: JSONConfig.plugins ? await getPlugins(JSONConfig.plugins) : void 0
  };
  return JSONConfig;
}

// src/utils/getCosmiConfig.ts
_chunkSYMMPQK7cjs.init_cjs_shims.call(void 0, );
var _bundlerequire = require('bundle-require');
var _cosmiconfig = require('cosmiconfig');
var tsLoader = async (configFile) => {
  const { mod } = await _bundlerequire.bundleRequire.call(void 0, {
    filepath: configFile,
    preserveTemporaryFile: false
  });
  return mod.default;
};
async function getCosmiConfig(moduleName, config) {
  const searchPlaces = [
    "package.json",
    `.${moduleName}rc`,
    `.${moduleName}rc.json`,
    `.${moduleName}rc.yaml`,
    `.${moduleName}rc.yml`,
    `.${moduleName}rc.ts`,
    `.${moduleName}rc.js`,
    `.${moduleName}rc.mjs`,
    `.${moduleName}rc.cjs`,
    `${moduleName}.config.ts`,
    `${moduleName}.config.js`,
    `${moduleName}.config.mjs`,
    `${moduleName}.config.cjs`
  ];
  const explorer = _cosmiconfig.cosmiconfig.call(void 0, moduleName, {
    cache: false,
    searchPlaces: [
      ...searchPlaces.map((searchPlace) => {
        return `.config/${searchPlace}`;
      }),
      ...searchPlaces.map((searchPlace) => {
        return `configs/${searchPlace}`;
      }),
      ...searchPlaces
    ],
    loaders: {
      ".ts": tsLoader
    }
  });
  const result = config ? await explorer.load(config) : await explorer.search();
  if (_optionalChain([result, 'optionalAccess', _8 => _8.isEmpty]) || !result || !result.config) {
    throw new Error("Config not defined, create a kubb.config.js or pass through your config with the option --config");
  }
  return result;
}

// src/utils/watcher.ts
_chunkSYMMPQK7cjs.init_cjs_shims.call(void 0, );
async function startWatcher(path2, cb) {
  const { watch } = await Promise.resolve().then(() => _interopRequireWildcard(require("chokidar")));
  const ignored = ["**/{.git,node_modules}/**"];
  const watcher = watch(path2, {
    ignorePermissionErrors: true,
    ignored
  });
  watcher.on("all", (type, file) => {
    _chunkSYMMPQK7cjs.spinner.succeed(_chunkSYMMPQK7cjs.p.yellow(_chunkSYMMPQK7cjs.p.bold(`Change detected: ${type} ${file}`)));
    _chunkSYMMPQK7cjs.spinner.spinner = "clock";
    try {
      cb(path2);
    } catch (e) {
      _chunkSYMMPQK7cjs.spinner.warn(_chunkSYMMPQK7cjs.p.red("Watcher failed"));
    }
  });
  return;
}

// src/commands/generate.ts

var args = {
  config: {
    type: "string",
    description: "Path to the Kubb config",
    alias: "c"
  },
  logLevel: {
    type: "string",
    description: "Info, silent or debug",
    alias: "l",
    default: _logger.LogLevel.info,
    valueHint: `${_logger.LogLevel.silent}|${_logger.LogLevel.info}|${_logger.LogLevel.debug}`
  },
  watch: {
    type: "boolean",
    description: "Watch mode based on the input file",
    alias: "w",
    default: false
  },
  bun: {
    type: "boolean",
    description: "Run Kubb with Bun",
    alias: "b",
    default: false
  },
  debug: {
    type: "boolean",
    description: "Override logLevel to debug",
    alias: "d",
    default: false
  },
  help: {
    type: "boolean",
    description: "Show help",
    alias: "h",
    default: false
  }
};
var command = _citty.defineCommand.call(void 0, {
  meta: {
    name: "generate",
    description: "[input] Generate files based on a 'kubb.config.ts' file"
  },
  args,
  setup() {
    _chunkSYMMPQK7cjs.spinner.start("\u{1F50D} Loading config");
  },
  async run({ args: args2 }) {
    const input = args2._[0];
    if (args2.help) {
      _citty.showUsage.call(void 0, command);
      return;
    }
    if (args2.debug) {
      args2.logLevel = _logger.LogLevel.debug;
    }
    if (args2.bun) {
      const command2 = process.argv.splice(2).filter((item) => item !== "--bun");
      await _chunkSYMMPQK7cjs.execa.call(void 0, "bkubb", command2, { stdout: process.stdout, stderr: process.stderr });
      return;
    }
    const result = await getCosmiConfig("kubb", args2.config);
    _chunkSYMMPQK7cjs.spinner.succeed(`\u{1F50D} Config loaded(${_chunkSYMMPQK7cjs.p.dim(_path2.default.relative(process.cwd(), result.filepath))})`);
    const config = await getConfig(result, args2);
    if (args2.watch) {
      if (Array.isArray(config)) {
        throw new Error("Cannot use watcher with multiple Configs(array)");
      }
      if (_core.isInputPath.call(void 0, config)) {
        return startWatcher([input || config.input.path], async (paths) => {
          await _chunkSYMMPQK7cjs.generate.call(void 0, { config, args: args2 });
          _chunkSYMMPQK7cjs.spinner.spinner = "simpleDotsScrolling";
          _chunkSYMMPQK7cjs.spinner.start(_chunkSYMMPQK7cjs.p.yellow(_chunkSYMMPQK7cjs.p.bold(`Watching for changes in ${paths.join(" and ")}`)));
        });
      }
    }
    if (Array.isArray(config)) {
      const promiseManager = new (0, _core.PromiseManager)();
      const promises = config.map((item) => () => _chunkSYMMPQK7cjs.generate.call(void 0, { input, config: item, args: args2 }));
      return promiseManager.run("seq", promises);
    }
    await _chunkSYMMPQK7cjs.generate.call(void 0, { input, config, args: args2 });
  }
});
var generate_default = command;


exports.default = generate_default;
