// @kubb/cli/dist/generate-NSWWKSNO.js
import {
  generate,
  spinner
} from "./chunk-M2AU2EGM.js";

// src/commands/generate.ts
import { LogLevel } from "@kubb/core/logger";
import { defineCommand, showUsage } from "citty";
import { execa } from "execa";
import c2 from "tinyrainbow";
import path from "node:path";

// src/utils/getConfig.ts
import { isPromise } from "@kubb/core/utils";

// src/utils/getPlugins.ts
import { PackageManager } from "@kubb/core";
function isJSONPlugins(plugins) {
  return !!plugins?.some((plugin) => {
    return Array.isArray(plugin) && typeof plugin?.at(0) === "string";
  });
}
function isObjectPlugins(plugins) {
  return plugins instanceof Object && !Array.isArray(plugins);
}
async function importPlugin(name, options) {
  const packageManager = new PackageManager(process.cwd());
  const importedPlugin = process.env.NODE_ENV === "test" ? await import(name) : await packageManager.import(name);
  return importedPlugin?.default ? importedPlugin.default(options) : importedPlugin(options);
}
function getPlugins(plugins) {
  if (isObjectPlugins(plugins)) {
    throw new Error("Object plugins are not supported anymore, best to use http://kubb.dev/guide/configure#json");
  }
  if (isJSONPlugins(plugins)) {
    const jsonPlugins = plugins;
    const promises = jsonPlugins.map((plugin) => {
      const [name, options = {}] = plugin;
      return importPlugin(name, options);
    });
    return Promise.all(promises);
  }
  return Promise.resolve(plugins);
}

// src/utils/getConfig.ts
async function getConfig(result, args2) {
  const config = result?.config;
  let kubbUserConfig = Promise.resolve(config);
  if (typeof config === "function") {
    const possiblePromise = config(args2);
    if (isPromise(possiblePromise)) {
      kubbUserConfig = possiblePromise;
    }
    kubbUserConfig = Promise.resolve(possiblePromise);
  }
  let JSONConfig = await kubbUserConfig;
  if (Array.isArray(JSONConfig)) {
    const promises = JSONConfig.map(async (item) => {
      return {
        ...item,
        plugins: item.plugins ? await getPlugins(item.plugins) : void 0
      };
    });
    return Promise.all(promises);
  }
  JSONConfig = {
    ...JSONConfig,
    plugins: JSONConfig.plugins ? await getPlugins(JSONConfig.plugins) : void 0
  };
  return JSONConfig;
}

// src/utils/getCosmiConfig.ts
import { bundleRequire } from "bundle-require";
import { cosmiconfig } from "cosmiconfig";
var tsLoader = async (configFile) => {
  const { mod } = await bundleRequire({
    filepath: configFile,
    preserveTemporaryFile: false
  });
  return mod.default;
};
async function getCosmiConfig(moduleName, config) {
  const searchPlaces = [
    "package.json",
    `.${moduleName}rc`,
    `.${moduleName}rc.json`,
    `.${moduleName}rc.yaml`,
    `.${moduleName}rc.yml`,
    `.${moduleName}rc.ts`,
    `.${moduleName}rc.js`,
    `.${moduleName}rc.mjs`,
    `.${moduleName}rc.cjs`,
    `${moduleName}.config.ts`,
    `${moduleName}.config.js`,
    `${moduleName}.config.mjs`,
    `${moduleName}.config.cjs`
  ];
  const explorer = cosmiconfig(moduleName, {
    cache: false,
    searchPlaces: [
      ...searchPlaces.map((searchPlace) => {
        return `.config/${searchPlace}`;
      }),
      ...searchPlaces.map((searchPlace) => {
        return `configs/${searchPlace}`;
      }),
      ...searchPlaces
    ],
    loaders: {
      ".ts": tsLoader
    }
  });
  const result = config ? await explorer.load(config) : await explorer.search();
  if (result?.isEmpty || !result || !result.config) {
    throw new Error("Config not defined, create a kubb.config.js or pass through your config with the option --config");
  }
  return result;
}

// src/utils/watcher.ts
import c from "tinyrainbow";
async function startWatcher(path2, cb) {
  const { watch } = await import("chokidar");
  const ignored = ["**/{.git,node_modules}/**"];
  const watcher = watch(path2, {
    ignorePermissionErrors: true,
    ignored
  });
  watcher.on("all", (type, file) => {
    spinner.succeed(c.yellow(c.bold(`Change detected: ${type} ${file}`)));
    spinner.spinner = "clock";
    try {
      cb(path2);
    } catch (e) {
      spinner.warn(c.red("Watcher failed"));
    }
  });
  return;
}

// src/commands/generate.ts
import { PromiseManager, isInputPath } from "@kubb/core";
var args = {
  config: {
    type: "string",
    description: "Path to the Kubb config",
    alias: "c"
  },
  logLevel: {
    type: "string",
    description: "Info, silent or debug",
    alias: "l",
    default: LogLevel.info,
    valueHint: `${LogLevel.silent}|${LogLevel.info}|${LogLevel.debug}`
  },
  watch: {
    type: "boolean",
    description: "Watch mode based on the input file",
    alias: "w",
    default: false
  },
  bun: {
    type: "boolean",
    description: "Run Kubb with Bun",
    alias: "b",
    default: false
  },
  debug: {
    type: "boolean",
    description: "Override logLevel to debug",
    alias: "d",
    default: false
  },
  help: {
    type: "boolean",
    description: "Show help",
    alias: "h",
    default: false
  }
};
var command = defineCommand({
  meta: {
    name: "generate",
    description: "[input] Generate files based on a 'kubb.config.ts' file"
  },
  args,
  setup() {
    spinner.start("\u{1F50D} Loading config");
  },
  async run({ args: args2 }) {
    const input = args2._[0];
    if (args2.help) {
      showUsage(command);
      return;
    }
    if (args2.debug) {
      args2.logLevel = LogLevel.debug;
    }
    if (args2.bun) {
      const command2 = process.argv.splice(2).filter((item) => item !== "--bun");
      await execa("bkubb", command2, { stdout: process.stdout, stderr: process.stderr });
      return;
    }
    const result = await getCosmiConfig("kubb", args2.config);
    spinner.succeed(`\u{1F50D} Config loaded(${c2.dim(path.relative(process.cwd(), result.filepath))})`);
    const config = await getConfig(result, args2);
    if (args2.watch) {
      if (Array.isArray(config)) {
        throw new Error("Cannot use watcher with multiple Configs(array)");
      }
      if (isInputPath(config)) {
        return startWatcher([input || config.input.path], async (paths) => {
          await generate({ config, args: args2 });
          spinner.spinner = "simpleDotsScrolling";
          spinner.start(c2.yellow(c2.bold(`Watching for changes in ${paths.join(" and ")}`)));
        });
      }
    }
    if (Array.isArray(config)) {
      const promiseManager = new PromiseManager();
      const promises = config.map((item) => () => generate({ input, config: item, args: args2 }));
      return promiseManager.run("seq", promises);
    }
    await generate({ input, config, args: args2 });
  }
});
var generate_default = command;
export {
  generate_default as default
};
