// @kubb/swagger-zod/dist/chunk-XBJLKQ6J.js
// src/components/Schema.tsx
import { Oas as Oas4 } from "@kubb/plugin-oas/components";
import { Const as Const2, File as File3, Type, useApp as useApp3, useFile } from "@kubb/react";
import { pluginTsName as pluginTsName2 } from "@kubb/swagger-ts";
import transformers3 from "@kubb/core/transformers";
import { isKeyword as isKeyword2, schemaKeywords as schemaKeywords3 } from "@kubb/plugin-oas";
import { useSchema } from "@kubb/plugin-oas/hooks";

// src/parser/index.ts
import transformers from "@kubb/core/transformers";
import { isKeyword, schemaKeywords } from "@kubb/plugin-oas";
var zodKeywordMapper = {
  any: () => "z.any()",
  unknown: () => "z.unknown()",
  number: (coercion, min, max) => {
    return [coercion ? "z.coerce.number()" : "z.number()", min !== void 0 ? `.min(${min})` : void 0, max !== void 0 ? `.max(${max})` : void 0].filter(Boolean).join("");
  },
  integer: (coercion, min, max) => {
    return [
      coercion ? "z.coerce.number().int()" : "z.number().int()",
      min !== void 0 ? `.min(${min})` : void 0,
      max !== void 0 ? `.max(${max})` : void 0
    ].filter(Boolean).join("");
  },
  object: (value) => `z.object({${value}})`,
  string: (coercion, min, max) => {
    return [coercion ? "z.coerce.string()" : "z.string()", min !== void 0 ? `.min(${min})` : void 0, max !== void 0 ? `.max(${max})` : void 0].filter(Boolean).join("");
  },
  boolean: () => "z.boolean()",
  undefined: () => "z.undefined()",
  nullable: () => ".nullable()",
  null: () => "z.null()",
  nullish: () => ".nullish()",
  array: (items = [], min, max) => {
    return [`z.array(${items?.join("")})`, min !== void 0 ? `.min(${min})` : void 0, max !== void 0 ? `.max(${max})` : void 0].filter(Boolean).join("");
  },
  tuple: (items = []) => `z.tuple([${items?.join(", ")}])`,
  enum: (items = []) => `z.enum([${items?.join(", ")}])`,
  union: (items = []) => `z.union([${items?.join(", ")}])`,
  const: (value) => `z.literal(${value ?? ""})`,
  /**
   * ISO 8601
   */
  datetime: (offset = false, local = false) => {
    if (offset) {
      return `z.string().datetime({ offset: ${offset} })`;
    }
    if (local) {
      return `z.string().datetime({ local: ${local} })`;
    }
    return "z.string().datetime()";
  },
  /**
   * Type `'date'` Date
   * Type `'string'` ISO date format (YYYY-MM-DD)
   * @default ISO date format (YYYY-MM-DD)
   */
  date: (type = "string", coercion) => {
    if (type === "string") {
      return "z.string().date()";
    }
    if (coercion) {
      return "z.coerce.date()";
    }
    return "z.date()";
  },
  /**
   * Type `'date'` Date
   * Type `'string'` ISO time format (HH:mm:ss[.SSSSSS])
   * @default ISO time format (HH:mm:ss[.SSSSSS])
   */
  time: (type = "string", coercion) => {
    if (type === "string") {
      return "z.string().time()";
    }
    if (coercion) {
      return "z.coerce.date()";
    }
    return "z.date()";
  },
  uuid: () => ".uuid()",
  url: () => ".url()",
  strict: () => ".strict()",
  default: (value) => `.default(${value ?? ""})`,
  and: (items = []) => items?.map((item) => `.and(${item})`).join(""),
  describe: (value = "") => `.describe(${value})`,
  min: (value) => `.min(${value ?? ""})`,
  max: (value) => `.max(${value ?? ""})`,
  optional: () => ".optional()",
  matches: (value = "") => `.regex(${value})`,
  email: () => ".email()",
  firstName: void 0,
  lastName: void 0,
  password: void 0,
  phone: void 0,
  readOnly: void 0,
  writeOnly: void 0,
  ref: (value) => value ? `z.lazy(() => ${value})` : void 0,
  blob: () => "z.string()",
  deprecated: void 0,
  example: void 0,
  schema: void 0,
  catchall: (value) => value ? `.catchall(${value})` : void 0,
  name: void 0
};
function sort(items) {
  const order = [
    schemaKeywords.string,
    schemaKeywords.datetime,
    schemaKeywords.date,
    schemaKeywords.time,
    schemaKeywords.tuple,
    schemaKeywords.number,
    schemaKeywords.object,
    schemaKeywords.enum,
    schemaKeywords.url,
    schemaKeywords.email,
    schemaKeywords.firstName,
    schemaKeywords.lastName,
    schemaKeywords.password,
    schemaKeywords.matches,
    schemaKeywords.uuid,
    schemaKeywords.min,
    schemaKeywords.max,
    schemaKeywords.default,
    schemaKeywords.describe,
    schemaKeywords.optional,
    schemaKeywords.nullable,
    schemaKeywords.nullish,
    schemaKeywords.null
  ];
  if (!items) {
    return [];
  }
  return transformers.orderBy(items, [(v) => order.indexOf(v.keyword)], ["asc"]);
}
var shouldCoerce = (coercion, type) => {
  if (coercion === void 0) {
    return false;
  }
  if (typeof coercion === "boolean") {
    return coercion;
  }
  return !!coercion[type];
};
function parse(parent, current, options) {
  const value = zodKeywordMapper[current.keyword];
  if (!value) {
    return void 0;
  }
  if (isKeyword(current, schemaKeywords.union)) {
    if (Array.isArray(current.args) && current.args.length === 1) {
      return parse(parent, current.args[0], options);
    }
    if (Array.isArray(current.args) && !current.args.length) {
      return "";
    }
    return zodKeywordMapper.union(
      sort(current.args).map((schema) => parse(current, schema, options)).filter(Boolean)
    );
  }
  if (isKeyword(current, schemaKeywords.and)) {
    const items = sort(current.args).filter((schema) => {
      return ![schemaKeywords.optional, schemaKeywords.describe].includes(schema.keyword);
    }).map((schema) => parse(current, schema, options)).filter(Boolean);
    return `${items.slice(0, 1)}${zodKeywordMapper.and(items.slice(1))}`;
  }
  if (isKeyword(current, schemaKeywords.array)) {
    return zodKeywordMapper.array(
      sort(current.args.items).map((schemas) => parse(current, schemas, options)).filter(Boolean),
      current.args.min,
      current.args.max
    );
  }
  if (isKeyword(current, schemaKeywords.enum)) {
    if (current.args.asConst) {
      if (current.args.items.length === 1) {
        return parse(
          current,
          {
            keyword: schemaKeywords.const,
            args: current.args.items[0]
          },
          options
        );
      }
      return zodKeywordMapper.union(
        current.args.items.map((schema) => {
          return parse(
            current,
            {
              keyword: schemaKeywords.const,
              args: schema
            },
            options
          );
        }).filter(Boolean)
      );
    }
    return zodKeywordMapper.enum(
      current.args.items.map((schema) => {
        if (schema.format === "boolean") {
          return transformers.stringify(schema.value);
        }
        if (schema.format === "number") {
          return transformers.stringify(schema.value);
        }
        return transformers.stringify(schema.value);
      })
    );
  }
  if (isKeyword(current, schemaKeywords.ref)) {
    return zodKeywordMapper.ref(current.args?.name);
  }
  if (isKeyword(current, schemaKeywords.object)) {
    const properties = Object.entries(current.args?.properties || {}).filter((item) => {
      const schema = item[1];
      return schema && typeof schema.map === "function";
    }).map(([name, schemas]) => {
      const nameSchema = schemas.find((schema) => schema.keyword === schemaKeywords.name);
      const mappedName = nameSchema?.args || name;
      if (options.mapper?.[mappedName]) {
        return `"${name}": ${options.mapper?.[mappedName]}`;
      }
      return `"${name}": ${sort(schemas).map((schema, array) => {
        return parse(current, schema, options);
      }).filter(Boolean).join("")}`;
    }).join(",");
    const additionalProperties = current.args?.additionalProperties?.length ? current.args.additionalProperties.map((schema) => parse(current, schema, options)).filter(Boolean).join("") : void 0;
    const text = [
      zodKeywordMapper.object(properties),
      current.args?.strict ? zodKeywordMapper.strict() : void 0,
      additionalProperties ? zodKeywordMapper.catchall(additionalProperties) : void 0
    ].filter(Boolean);
    return text.join("");
  }
  if (isKeyword(current, schemaKeywords.tuple)) {
    return zodKeywordMapper.tuple(current.args.items.map((schema) => parse(current, schema, options)).filter(Boolean));
  }
  if (isKeyword(current, schemaKeywords.const)) {
    if (current.args.format === "number" && current.args.value !== void 0) {
      return zodKeywordMapper.const(Number.parseInt(current.args.value?.toString()));
    }
    if (current.args.format === "boolean" && current.args.value !== void 0) {
      return zodKeywordMapper.const(current.args.value);
    }
    return zodKeywordMapper.const(transformers.stringify(current.args.value));
  }
  if (isKeyword(current, schemaKeywords.matches)) {
    if (current.args) {
      return zodKeywordMapper.matches(transformers.toRegExpString(current.args));
    }
  }
  if (isKeyword(current, schemaKeywords.default)) {
    if (current.args) {
      return zodKeywordMapper.default(current.args);
    }
  }
  if (isKeyword(current, schemaKeywords.describe)) {
    if (current.args) {
      return zodKeywordMapper.describe(transformers.stringify(current.args.toString()));
    }
  }
  if (isKeyword(current, schemaKeywords.string)) {
    return zodKeywordMapper.string(shouldCoerce(options.coercion, "strings"));
  }
  if (isKeyword(current, schemaKeywords.number)) {
    return zodKeywordMapper.number(shouldCoerce(options.coercion, "numbers"));
  }
  if (isKeyword(current, schemaKeywords.integer)) {
    return zodKeywordMapper.integer(shouldCoerce(options.coercion, "numbers"));
  }
  if (isKeyword(current, schemaKeywords.min)) {
    return zodKeywordMapper.min(current.args);
  }
  if (isKeyword(current, schemaKeywords.max)) {
    return zodKeywordMapper.max(current.args);
  }
  if (isKeyword(current, schemaKeywords.datetime)) {
    return zodKeywordMapper.datetime(current.args.offset, current.args.local);
  }
  if (isKeyword(current, schemaKeywords.date)) {
    return zodKeywordMapper.date(current.args.type, shouldCoerce(options.coercion, "dates"));
  }
  if (isKeyword(current, schemaKeywords.time)) {
    return zodKeywordMapper.time(current.args.type, shouldCoerce(options.coercion, "dates"));
  }
  if (current.keyword in zodKeywordMapper && "args" in current) {
    const value2 = zodKeywordMapper[current.keyword];
    return value2(current.args);
  }
  if (current.keyword in zodKeywordMapper) {
    return value();
  }
  return void 0;
}

// src/plugin.ts
import path from "node:path";
import { FileManager, PluginManager, createPlugin } from "@kubb/core";
import { camelCase, pascalCase } from "@kubb/core/transformers";
import { renderTemplate } from "@kubb/core/utils";
import { pluginOasName } from "@kubb/plugin-oas";
import { getGroupedByTagFiles } from "@kubb/plugin-oas/utils";
import { pluginTsName } from "@kubb/swagger-ts";

// src/OperationGenerator.tsx
import { OperationGenerator as Generator2 } from "@kubb/plugin-oas";
import { Oas as Oas3 } from "@kubb/plugin-oas/components";
import { App as App2, createRoot as createRoot2 } from "@kubb/react";

// src/components/OperationSchema.tsx
import { schemaKeywords as schemaKeywords2 } from "@kubb/plugin-oas";
import { Oas as Oas2 } from "@kubb/plugin-oas/components";
import { useOas, useOperation, useOperationManager } from "@kubb/plugin-oas/hooks";
import { File, Parser, useApp } from "@kubb/react";

// src/SchemaGenerator.tsx
import { SchemaGenerator as Generator } from "@kubb/plugin-oas";
import { Oas } from "@kubb/plugin-oas/components";
import { App, createRoot } from "@kubb/react";
import { jsx } from "@kubb/react/jsx-runtime";
var SchemaGenerator = class extends Generator {
  async schema(name, schema, options) {
    const { oas, pluginManager, plugin, mode, output } = this.context;
    const root = createRoot({
      logger: pluginManager.logger
    });
    const tree = this.parse({ schema, name });
    root.render(
      /* @__PURE__ */ jsx(App, { pluginManager, plugin: { ...plugin, options }, mode, children: /* @__PURE__ */ jsx(Oas, { oas, children: /* @__PURE__ */ jsx(Oas.Schema, { name, value: schema, tree, children: /* @__PURE__ */ jsx(Schema.File, {}) }) }) })
    );
    return root.files;
  }
};

// src/components/OperationSchema.tsx
import { jsx as jsx2, jsxs } from "@kubb/react/jsx-runtime";
function OperationSchema({ description, keysToOmit }) {
  return /* @__PURE__ */ jsx2(Schema, { keysToOmit, withTypeAnnotation: false, description });
}
OperationSchema.File = function({}) {
  const { pluginManager, plugin, mode } = useApp();
  const { getSchemas, getFile } = useOperationManager();
  const oas = useOas();
  const operation = useOperation();
  const file = getFile(operation);
  const schemas = getSchemas(operation);
  const generator = new SchemaGenerator(plugin.options, {
    oas,
    plugin,
    pluginManager,
    mode,
    override: plugin.options.override
  });
  const items = [schemas.pathParams, schemas.queryParams, schemas.headerParams, schemas.statusCodes, schemas.request, schemas.response].flat().filter(Boolean);
  const mapItem = ({ name, schema, description, keysToOmit, ...options }, i) => {
    const required = Array.isArray(schema?.required) ? !!schema.required.length : !!schema?.required;
    const optional = !required && !!name.includes("Params");
    const tree = generator.parse({ schema, name });
    return /* @__PURE__ */ jsxs(Oas2.Schema, { name, value: schema, tree: [...tree, optional ? { keyword: schemaKeywords2.optional } : void 0].filter(Boolean), children: [
      mode === "split" && /* @__PURE__ */ jsx2(Oas2.Schema.Imports, { extName: plugin.options.extName }),
      /* @__PURE__ */ jsx2(File.Source, { children: /* @__PURE__ */ jsx2(OperationSchema, { description, keysToOmit }) })
    ] }, i);
  };
  return /* @__PURE__ */ jsx2(Parser, { language: "typescript", children: /* @__PURE__ */ jsxs(File, { baseName: file.baseName, path: file.path, meta: file.meta, children: [
    /* @__PURE__ */ jsx2(File.Import, { name: ["z"], path: plugin.options.importPath }),
    items.map(mapItem)
  ] }) });
};

// src/components/Operations.tsx
import { useOperationManager as useOperationManager2, useOperations } from "@kubb/plugin-oas/hooks";
import { Const, File as File2, Parser as Parser2, useApp as useApp2 } from "@kubb/react";
import transformers2 from "@kubb/core/transformers";
import { Fragment, jsx as jsx3, jsxs as jsxs2 } from "@kubb/react/jsx-runtime";
function Template({ operationsName, pathsName, operations }) {
  const { groupSchemasByName } = useOperationManager2();
  const transformedOperations = operations.map((operation) => ({ operation, data: groupSchemasByName(operation, { type: "function" }) }));
  const operationsJSON = transformedOperations.reduce(
    (prev, acc) => {
      prev[`"${acc.operation.getOperationId()}"`] = acc.data;
      return prev;
    },
    {}
  );
  const pathsJSON = transformedOperations.reduce(
    (prev, acc) => {
      prev[`"${acc.operation.path}"`] = {
        ...prev[`"${acc.operation.path}"`] || {},
        [acc.operation.method]: `operations["${acc.operation.getOperationId()}"]`
      };
      return prev;
    },
    {}
  );
  return /* @__PURE__ */ jsxs2(Fragment, { children: [
    /* @__PURE__ */ jsx3(Const, { export: true, name: operationsName, asConst: true, children: `{${transformers2.stringifyObject(operationsJSON)}}` }),
    /* @__PURE__ */ jsx3(Const, { export: true, name: pathsName, asConst: true, children: `{${transformers2.stringifyObject(pathsJSON)}}` })
  ] });
}
function RootTemplate({ children }) {
  const {
    mode,
    pluginManager,
    plugin: {
      key: pluginKey,
      options: { extName }
    }
  } = useApp2();
  const { getFile } = useOperationManager2();
  const operations = useOperations();
  const { groupSchemasByName } = useOperationManager2();
  const transformedOperations = operations.map((operation) => ({ operation, data: groupSchemasByName(operation, { type: "function" }) }));
  const file = pluginManager.getFile({ name: "operations", extName: ".ts", pluginKey });
  const imports = Object.entries(transformedOperations).map(([_key, { data, operation }], index) => {
    const names = [data.request, ...Object.values(data.responses), ...Object.values(data.parameters)].filter(Boolean);
    return /* @__PURE__ */ jsx3(File2.Import, { extName, name: names, root: file.path, path: getFile(operation).path }, index);
  }).filter(Boolean);
  return /* @__PURE__ */ jsx3(Parser2, { language: "typescript", children: /* @__PURE__ */ jsxs2(File2, { baseName: file.baseName, path: file.path, meta: file.meta, exportable: false, children: [
    mode === "split" && imports,
    /* @__PURE__ */ jsx3(File2.Source, { children })
  ] }) });
}
var defaultTemplates = { default: Template, root: RootTemplate };
function Operations({ Template: Template2 = defaultTemplates.default }) {
  const operations = useOperations();
  return /* @__PURE__ */ jsx3(Template2, { operationsName: "operations", pathsName: "paths", operations });
}
Operations.File = function(props) {
  const templates = { ...defaultTemplates, ...props.templates };
  const Template2 = templates.default;
  const RootTemplate2 = templates.root;
  return /* @__PURE__ */ jsx3(RootTemplate2, { children: /* @__PURE__ */ jsx3(Operations, { Template: Template2 }) });
};
Operations.templates = defaultTemplates;

// src/OperationGenerator.tsx
import { jsx as jsx4 } from "@kubb/react/jsx-runtime";
var OperationGenerator = class extends Generator2 {
  async all(operations, _operationsByMethod) {
    const { pluginManager, oas, plugin, mode } = this.context;
    const root = createRoot2({
      logger: pluginManager.logger
    });
    const templates = {
      operations: Operations.templates,
      ...this.options.templates
    };
    root.render(
      /* @__PURE__ */ jsx4(App2, { pluginManager, plugin, mode, children: /* @__PURE__ */ jsx4(Oas3, { oas, operations, generator: this, children: templates.operations && /* @__PURE__ */ jsx4(Operations.File, { templates: templates.operations }) }) })
    );
    return root.files;
  }
  async operation(operation, options) {
    const { oas, pluginManager, plugin, mode } = this.context;
    const root = createRoot2({
      logger: pluginManager.logger
    });
    root.render(
      /* @__PURE__ */ jsx4(App2, { pluginManager, plugin: { ...plugin, options }, mode, children: /* @__PURE__ */ jsx4(Oas3, { oas, operations: [operation], generator: this, children: /* @__PURE__ */ jsx4(Oas3.Operation, { operation, children: /* @__PURE__ */ jsx4(OperationSchema.File, {}) }) }) })
    );
    return root.files;
  }
};

// src/plugin.ts
var pluginZodName = "plugin-zod";
var pluginZod = createPlugin((options) => {
  const {
    output = { path: "zod" },
    group,
    exclude = [],
    include,
    override = [],
    transformers: transformers4 = {},
    dateType = "string",
    unknownType = "any",
    typed = false,
    typedSchema = false,
    mapper = {},
    templates,
    importPath = "zod",
    coercion = false
  } = options;
  const template = group?.output ? group.output : `${output.path}/{{tag}}Controller`;
  return {
    name: pluginZodName,
    options: {
      extName: output.extName,
      transformers: transformers4,
      include,
      exclude,
      override,
      typed,
      typedSchema,
      dateType,
      unknownType,
      mapper,
      importPath,
      coercion,
      templates: {
        operations: Operations.templates,
        ...templates
      }
    },
    pre: [pluginOasName, typed ? pluginTsName : void 0].filter(Boolean),
    resolvePath(baseName, pathMode, options2) {
      const root = path.resolve(this.config.root, this.config.output.path);
      const mode = pathMode ?? FileManager.getMode(path.resolve(root, output.path));
      if (mode === "single") {
        return path.resolve(root, output.path);
      }
      if (options2?.tag && group?.type === "tag") {
        const tag = camelCase(options2.tag);
        return path.resolve(root, renderTemplate(template, { tag }), baseName);
      }
      return path.resolve(root, output.path, baseName);
    },
    resolveName(name, type) {
      let resolvedName = camelCase(name, {
        suffix: type ? "schema" : void 0,
        isFile: type === "file"
      });
      if (type === "type") {
        resolvedName = pascalCase(resolvedName);
      }
      if (type) {
        return transformers4?.name?.(resolvedName, type) || resolvedName;
      }
      return resolvedName;
    },
    async writeFile(path2, source) {
      if (!path2.endsWith(".ts") || !source) {
        return;
      }
      return this.fileManager.write(path2, source, { sanity: false });
    },
    async buildStart() {
      const [swaggerPlugin] = PluginManager.getDependedPlugins(this.plugins, [pluginOasName]);
      const oas = await swaggerPlugin.api.getOas();
      const root = path.resolve(this.config.root, this.config.output.path);
      const mode = FileManager.getMode(path.resolve(root, output.path));
      const schemaGenerator = new SchemaGenerator(this.plugin.options, {
        oas,
        pluginManager: this.pluginManager,
        plugin: this.plugin,
        contentType: swaggerPlugin.api.contentType,
        include: void 0,
        override,
        mode,
        output: output.path
      });
      const schemaFiles = await schemaGenerator.build();
      await this.addFile(...schemaFiles);
      const operationGenerator = new OperationGenerator(this.plugin.options, {
        oas,
        pluginManager: this.pluginManager,
        plugin: this.plugin,
        contentType: swaggerPlugin.api.contentType,
        exclude,
        include,
        override,
        mode
      });
      const operationFiles = await operationGenerator.build();
      await this.addFile(...operationFiles);
    },
    async buildEnd() {
      if (this.config.output.write === false) {
        return;
      }
      const root = path.resolve(this.config.root, this.config.output.path);
      if (group?.type === "tag") {
        const rootFiles = await getGroupedByTagFiles({
          logger: this.logger,
          files: this.fileManager.files,
          plugin: this.plugin,
          template,
          exportAs: group.exportAs || "{{tag}}Schemas",
          root,
          output
        });
        await this.addFile(...rootFiles);
      }
      await this.fileManager.addIndexes({
        root,
        output,
        meta: { pluginKey: this.plugin.key },
        logger: this.logger
      });
    }
  };
});

// src/components/Schema.tsx
import { Fragment as Fragment2, jsx as jsx5, jsxs as jsxs3 } from "@kubb/react/jsx-runtime";
function Schema(props) {
  const { keysToOmit, withTypeAnnotation, description } = props;
  const { tree, name } = useSchema();
  const {
    pluginManager,
    plugin: {
      options: { mapper, typedSchema, coercion }
    }
  } = useApp3();
  const resolvedName = pluginManager.resolveName({
    name,
    pluginKey: [pluginZodName],
    type: "function"
  });
  const resolvedTypeName = pluginManager.resolveName({
    name,
    pluginKey: [pluginZodName],
    type: "type"
  });
  const typeName = pluginManager.resolveName({
    name,
    pluginKey: [pluginTsName2],
    type: "type"
  });
  if (!tree.length) {
    return /* @__PURE__ */ jsx5(
      Const2,
      {
        name: resolvedName,
        export: true,
        JSDoc: {
          comments: [description ? `@description ${transformers3.jsStringEscape(description)}` : void 0].filter(Boolean)
        },
        children: "undefined"
      }
    );
  }
  const hasTuple = tree.some((item) => isKeyword2(item, schemaKeywords3.tuple));
  const output = sort(tree).filter((item) => {
    if (hasTuple && (isKeyword2(item, schemaKeywords3.min) || isKeyword2(item, schemaKeywords3.max))) {
      return false;
    }
    return true;
  }).map((item) => parse(void 0, item, { name, typeName, description, mapper, coercion, keysToOmit })).filter(Boolean).join("");
  const suffix = output.endsWith(".nullable()") ? ".unwrap().and" : ".and";
  return /* @__PURE__ */ jsxs3(Fragment2, { children: [
    /* @__PURE__ */ jsx5(
      Const2,
      {
        export: true,
        name: resolvedName,
        JSDoc: {
          comments: [description ? `@description ${transformers3.jsStringEscape(description)}` : void 0].filter(Boolean)
        },
        children: [
          output,
          keysToOmit?.length ? `${suffix}(z.object({ ${keysToOmit.map((key) => `${key}: z.never()`).join(",")} }))` : void 0,
          withTypeAnnotation && typeName ? ` as z.ZodType<${typeName}>` : ""
        ].filter(Boolean).join("") || ""
      }
    ),
    typedSchema && /* @__PURE__ */ jsx5(Type, { export: true, name: resolvedTypeName, children: `z.infer<typeof ${resolvedName}>` })
  ] });
}
Schema.File = function({}) {
  const {
    pluginManager,
    plugin: {
      options: { typed }
    }
  } = useApp3();
  const { tree, schema } = useSchema();
  const withData = tree.some(
    (schema2) => schema2.keyword === schemaKeywords3.array || schema2.keyword === schemaKeywords3.and || schema2.keyword === schemaKeywords3.object || schema2.keyword === schemaKeywords3.union || schema2.keyword === schemaKeywords3.tuple
  );
  const withTypeAnnotation = !!typed;
  return /* @__PURE__ */ jsxs3(Oas4.Schema.File, { output: pluginManager.config.output.path, children: [
    /* @__PURE__ */ jsx5(Schema.Imports, {}),
    /* @__PURE__ */ jsx5(File3.Source, { children: /* @__PURE__ */ jsx5(Schema, { withTypeAnnotation, description: schema?.description }) })
  ] });
};
Schema.Imports = () => {
  const {
    pluginManager,
    plugin: {
      options: { typed, importPath }
    }
  } = useApp3();
  const { path: root } = useFile();
  const { name, tree, schema } = useSchema();
  const typeName = pluginManager.resolveName({
    name,
    pluginKey: [pluginTsName2],
    type: "type"
  });
  const typeFileName = pluginManager.resolveName({
    name,
    pluginKey: [pluginTsName2],
    type: "file"
  });
  const typePath = pluginManager.resolvePath({
    baseName: typeFileName,
    pluginKey: [pluginTsName2]
  });
  const withTypeAnnotation = !!typed;
  return /* @__PURE__ */ jsxs3(Fragment2, { children: [
    /* @__PURE__ */ jsx5(File3.Import, { name: ["z"], path: importPath }),
    withTypeAnnotation && typeName && typePath && /* @__PURE__ */ jsx5(File3.Import, { isTypeOnly: true, root, path: typePath, name: [typeName] })
  ] });
};

export {
  Schema,
  OperationSchema,
  Operations,
  pluginZodName,
  pluginZod
};
