// @kubb/swagger-zod/dist/chunk-J3CIE7RB.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; } function _nullishCoalesce(lhs, rhsFn) { if (lhs != null) { return lhs; } else { return rhsFn(); } } function _optionalChain(ops) { let lastAccessLHS = undefined; let value = ops[0]; let i = 1; while (i < ops.length) { const op = ops[i]; const fn = ops[i + 1]; i += 2; if ((op === 'optionalAccess' || op === 'optionalCall') && value == null) { return undefined; } if (op === 'access' || op === 'optionalAccess') { lastAccessLHS = value; value = fn(value); } else if (op === 'call' || op === 'optionalCall') { value = fn((...args) => value.call(lastAccessLHS, ...args)); lastAccessLHS = undefined; } } return value; }// src/components/Schema.tsx
var _components = require('@kubb/plugin-oas/components');
var _react = require('@kubb/react');
var _swaggerts = require('@kubb/swagger-ts');
var _transformers = require('@kubb/core/transformers'); var _transformers2 = _interopRequireDefault(_transformers);
var _pluginoas = require('@kubb/plugin-oas');
var _hooks = require('@kubb/plugin-oas/hooks');

// src/parser/index.ts


var zodKeywordMapper = {
  any: () => "z.any()",
  unknown: () => "z.unknown()",
  number: (coercion, min, max) => {
    return [coercion ? "z.coerce.number()" : "z.number()", min !== void 0 ? `.min(${min})` : void 0, max !== void 0 ? `.max(${max})` : void 0].filter(Boolean).join("");
  },
  integer: (coercion, min, max) => {
    return [
      coercion ? "z.coerce.number().int()" : "z.number().int()",
      min !== void 0 ? `.min(${min})` : void 0,
      max !== void 0 ? `.max(${max})` : void 0
    ].filter(Boolean).join("");
  },
  object: (value) => `z.object({${value}})`,
  string: (coercion, min, max) => {
    return [coercion ? "z.coerce.string()" : "z.string()", min !== void 0 ? `.min(${min})` : void 0, max !== void 0 ? `.max(${max})` : void 0].filter(Boolean).join("");
  },
  boolean: () => "z.boolean()",
  undefined: () => "z.undefined()",
  nullable: () => ".nullable()",
  null: () => "z.null()",
  nullish: () => ".nullish()",
  array: (items = [], min, max) => {
    return [`z.array(${_optionalChain([items, 'optionalAccess', _ => _.join, 'call', _2 => _2("")])})`, min !== void 0 ? `.min(${min})` : void 0, max !== void 0 ? `.max(${max})` : void 0].filter(Boolean).join("");
  },
  tuple: (items = []) => `z.tuple([${_optionalChain([items, 'optionalAccess', _3 => _3.join, 'call', _4 => _4(", ")])}])`,
  enum: (items = []) => `z.enum([${_optionalChain([items, 'optionalAccess', _5 => _5.join, 'call', _6 => _6(", ")])}])`,
  union: (items = []) => `z.union([${_optionalChain([items, 'optionalAccess', _7 => _7.join, 'call', _8 => _8(", ")])}])`,
  const: (value) => `z.literal(${_nullishCoalesce(value, () => ( ""))})`,
  /**
   * ISO 8601
   */
  datetime: (offset = false, local = false) => {
    if (offset) {
      return `z.string().datetime({ offset: ${offset} })`;
    }
    if (local) {
      return `z.string().datetime({ local: ${local} })`;
    }
    return "z.string().datetime()";
  },
  /**
   * Type `'date'` Date
   * Type `'string'` ISO date format (YYYY-MM-DD)
   * @default ISO date format (YYYY-MM-DD)
   */
  date: (type = "string", coercion) => {
    if (type === "string") {
      return "z.string().date()";
    }
    if (coercion) {
      return "z.coerce.date()";
    }
    return "z.date()";
  },
  /**
   * Type `'date'` Date
   * Type `'string'` ISO time format (HH:mm:ss[.SSSSSS])
   * @default ISO time format (HH:mm:ss[.SSSSSS])
   */
  time: (type = "string", coercion) => {
    if (type === "string") {
      return "z.string().time()";
    }
    if (coercion) {
      return "z.coerce.date()";
    }
    return "z.date()";
  },
  uuid: () => ".uuid()",
  url: () => ".url()",
  strict: () => ".strict()",
  default: (value) => `.default(${_nullishCoalesce(value, () => ( ""))})`,
  and: (items = []) => _optionalChain([items, 'optionalAccess', _9 => _9.map, 'call', _10 => _10((item) => `.and(${item})`), 'access', _11 => _11.join, 'call', _12 => _12("")]),
  describe: (value = "") => `.describe(${value})`,
  min: (value) => `.min(${_nullishCoalesce(value, () => ( ""))})`,
  max: (value) => `.max(${_nullishCoalesce(value, () => ( ""))})`,
  optional: () => ".optional()",
  matches: (value = "") => `.regex(${value})`,
  email: () => ".email()",
  firstName: void 0,
  lastName: void 0,
  password: void 0,
  phone: void 0,
  readOnly: void 0,
  writeOnly: void 0,
  ref: (value) => value ? `z.lazy(() => ${value})` : void 0,
  blob: () => "z.string()",
  deprecated: void 0,
  example: void 0,
  schema: void 0,
  catchall: (value) => value ? `.catchall(${value})` : void 0,
  name: void 0
};
function sort(items) {
  const order = [
    _pluginoas.schemaKeywords.string,
    _pluginoas.schemaKeywords.datetime,
    _pluginoas.schemaKeywords.date,
    _pluginoas.schemaKeywords.time,
    _pluginoas.schemaKeywords.tuple,
    _pluginoas.schemaKeywords.number,
    _pluginoas.schemaKeywords.object,
    _pluginoas.schemaKeywords.enum,
    _pluginoas.schemaKeywords.url,
    _pluginoas.schemaKeywords.email,
    _pluginoas.schemaKeywords.firstName,
    _pluginoas.schemaKeywords.lastName,
    _pluginoas.schemaKeywords.password,
    _pluginoas.schemaKeywords.matches,
    _pluginoas.schemaKeywords.uuid,
    _pluginoas.schemaKeywords.min,
    _pluginoas.schemaKeywords.max,
    _pluginoas.schemaKeywords.default,
    _pluginoas.schemaKeywords.describe,
    _pluginoas.schemaKeywords.optional,
    _pluginoas.schemaKeywords.nullable,
    _pluginoas.schemaKeywords.nullish,
    _pluginoas.schemaKeywords.null
  ];
  if (!items) {
    return [];
  }
  return _transformers2.default.orderBy(items, [(v) => order.indexOf(v.keyword)], ["asc"]);
}
var shouldCoerce = (coercion, type) => {
  if (coercion === void 0) {
    return false;
  }
  if (typeof coercion === "boolean") {
    return coercion;
  }
  return !!coercion[type];
};
function parse(parent, current, options) {
  const value = zodKeywordMapper[current.keyword];
  if (!value) {
    return void 0;
  }
  if (_pluginoas.isKeyword.call(void 0, current, _pluginoas.schemaKeywords.union)) {
    if (Array.isArray(current.args) && current.args.length === 1) {
      return parse(parent, current.args[0], options);
    }
    if (Array.isArray(current.args) && !current.args.length) {
      return "";
    }
    return zodKeywordMapper.union(
      sort(current.args).map((schema) => parse(current, schema, options)).filter(Boolean)
    );
  }
  if (_pluginoas.isKeyword.call(void 0, current, _pluginoas.schemaKeywords.and)) {
    const items = sort(current.args).filter((schema) => {
      return ![_pluginoas.schemaKeywords.optional, _pluginoas.schemaKeywords.describe].includes(schema.keyword);
    }).map((schema) => parse(current, schema, options)).filter(Boolean);
    return `${items.slice(0, 1)}${zodKeywordMapper.and(items.slice(1))}`;
  }
  if (_pluginoas.isKeyword.call(void 0, current, _pluginoas.schemaKeywords.array)) {
    return zodKeywordMapper.array(
      sort(current.args.items).map((schemas) => parse(current, schemas, options)).filter(Boolean),
      current.args.min,
      current.args.max
    );
  }
  if (_pluginoas.isKeyword.call(void 0, current, _pluginoas.schemaKeywords.enum)) {
    if (current.args.asConst) {
      if (current.args.items.length === 1) {
        return parse(
          current,
          {
            keyword: _pluginoas.schemaKeywords.const,
            args: current.args.items[0]
          },
          options
        );
      }
      return zodKeywordMapper.union(
        current.args.items.map((schema) => {
          return parse(
            current,
            {
              keyword: _pluginoas.schemaKeywords.const,
              args: schema
            },
            options
          );
        }).filter(Boolean)
      );
    }
    return zodKeywordMapper.enum(
      current.args.items.map((schema) => {
        if (schema.format === "boolean") {
          return _transformers2.default.stringify(schema.value);
        }
        if (schema.format === "number") {
          return _transformers2.default.stringify(schema.value);
        }
        return _transformers2.default.stringify(schema.value);
      })
    );
  }
  if (_pluginoas.isKeyword.call(void 0, current, _pluginoas.schemaKeywords.ref)) {
    return zodKeywordMapper.ref(_optionalChain([current, 'access', _13 => _13.args, 'optionalAccess', _14 => _14.name]));
  }
  if (_pluginoas.isKeyword.call(void 0, current, _pluginoas.schemaKeywords.object)) {
    const properties = Object.entries(_optionalChain([current, 'access', _15 => _15.args, 'optionalAccess', _16 => _16.properties]) || {}).filter((item) => {
      const schema = item[1];
      return schema && typeof schema.map === "function";
    }).map(([name, schemas]) => {
      const nameSchema = schemas.find((schema) => schema.keyword === _pluginoas.schemaKeywords.name);
      const mappedName = _optionalChain([nameSchema, 'optionalAccess', _17 => _17.args]) || name;
      if (_optionalChain([options, 'access', _18 => _18.mapper, 'optionalAccess', _19 => _19[mappedName]])) {
        return `"${name}": ${_optionalChain([options, 'access', _20 => _20.mapper, 'optionalAccess', _21 => _21[mappedName]])}`;
      }
      return `"${name}": ${sort(schemas).map((schema, array) => {
        return parse(current, schema, options);
      }).filter(Boolean).join("")}`;
    }).join(",");
    const additionalProperties = _optionalChain([current, 'access', _22 => _22.args, 'optionalAccess', _23 => _23.additionalProperties, 'optionalAccess', _24 => _24.length]) ? current.args.additionalProperties.map((schema) => parse(current, schema, options)).filter(Boolean).join("") : void 0;
    const text = [
      zodKeywordMapper.object(properties),
      _optionalChain([current, 'access', _25 => _25.args, 'optionalAccess', _26 => _26.strict]) ? zodKeywordMapper.strict() : void 0,
      additionalProperties ? zodKeywordMapper.catchall(additionalProperties) : void 0
    ].filter(Boolean);
    return text.join("");
  }
  if (_pluginoas.isKeyword.call(void 0, current, _pluginoas.schemaKeywords.tuple)) {
    return zodKeywordMapper.tuple(current.args.items.map((schema) => parse(current, schema, options)).filter(Boolean));
  }
  if (_pluginoas.isKeyword.call(void 0, current, _pluginoas.schemaKeywords.const)) {
    if (current.args.format === "number" && current.args.value !== void 0) {
      return zodKeywordMapper.const(Number.parseInt(_optionalChain([current, 'access', _27 => _27.args, 'access', _28 => _28.value, 'optionalAccess', _29 => _29.toString, 'call', _30 => _30()])));
    }
    if (current.args.format === "boolean" && current.args.value !== void 0) {
      return zodKeywordMapper.const(current.args.value);
    }
    return zodKeywordMapper.const(_transformers2.default.stringify(current.args.value));
  }
  if (_pluginoas.isKeyword.call(void 0, current, _pluginoas.schemaKeywords.matches)) {
    if (current.args) {
      return zodKeywordMapper.matches(_transformers2.default.toRegExpString(current.args));
    }
  }
  if (_pluginoas.isKeyword.call(void 0, current, _pluginoas.schemaKeywords.default)) {
    if (current.args) {
      return zodKeywordMapper.default(current.args);
    }
  }
  if (_pluginoas.isKeyword.call(void 0, current, _pluginoas.schemaKeywords.describe)) {
    if (current.args) {
      return zodKeywordMapper.describe(_transformers2.default.stringify(current.args.toString()));
    }
  }
  if (_pluginoas.isKeyword.call(void 0, current, _pluginoas.schemaKeywords.string)) {
    return zodKeywordMapper.string(shouldCoerce(options.coercion, "strings"));
  }
  if (_pluginoas.isKeyword.call(void 0, current, _pluginoas.schemaKeywords.number)) {
    return zodKeywordMapper.number(shouldCoerce(options.coercion, "numbers"));
  }
  if (_pluginoas.isKeyword.call(void 0, current, _pluginoas.schemaKeywords.integer)) {
    return zodKeywordMapper.integer(shouldCoerce(options.coercion, "numbers"));
  }
  if (_pluginoas.isKeyword.call(void 0, current, _pluginoas.schemaKeywords.min)) {
    return zodKeywordMapper.min(current.args);
  }
  if (_pluginoas.isKeyword.call(void 0, current, _pluginoas.schemaKeywords.max)) {
    return zodKeywordMapper.max(current.args);
  }
  if (_pluginoas.isKeyword.call(void 0, current, _pluginoas.schemaKeywords.datetime)) {
    return zodKeywordMapper.datetime(current.args.offset, current.args.local);
  }
  if (_pluginoas.isKeyword.call(void 0, current, _pluginoas.schemaKeywords.date)) {
    return zodKeywordMapper.date(current.args.type, shouldCoerce(options.coercion, "dates"));
  }
  if (_pluginoas.isKeyword.call(void 0, current, _pluginoas.schemaKeywords.time)) {
    return zodKeywordMapper.time(current.args.type, shouldCoerce(options.coercion, "dates"));
  }
  if (current.keyword in zodKeywordMapper && "args" in current) {
    const value2 = zodKeywordMapper[current.keyword];
    return value2(current.args);
  }
  if (current.keyword in zodKeywordMapper) {
    return value();
  }
  return void 0;
}

// src/plugin.ts
var _path = require('path'); var _path2 = _interopRequireDefault(_path);
var _core = require('@kubb/core');

var _utils = require('@kubb/core/utils');

var _utils3 = require('@kubb/plugin-oas/utils');


// src/OperationGenerator.tsx




// src/components/OperationSchema.tsx





// src/SchemaGenerator.tsx



var _jsxruntime = require('@kubb/react/jsx-runtime');
var SchemaGenerator = class extends _pluginoas.SchemaGenerator {
  async schema(name, schema, options) {
    const { oas, pluginManager, plugin, mode, output } = this.context;
    const root = _react.createRoot.call(void 0, {
      logger: pluginManager.logger
    });
    const tree = this.parse({ schema, name });
    root.render(
      /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _react.App, { pluginManager, plugin: { ...plugin, options }, mode, children: /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _components.Oas, { oas, children: /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _components.Oas.Schema, { name, value: schema, tree, children: /* @__PURE__ */ _jsxruntime.jsx.call(void 0, Schema.File, {}) }) }) })
    );
    return root.files;
  }
};

// src/components/OperationSchema.tsx

function OperationSchema({ description, keysToOmit }) {
  return /* @__PURE__ */ _jsxruntime.jsx.call(void 0, Schema, { keysToOmit, withTypeAnnotation: false, description });
}
OperationSchema.File = function({}) {
  const { pluginManager, plugin, mode } = _react.useApp.call(void 0, );
  const { getSchemas, getFile } = _hooks.useOperationManager.call(void 0, );
  const oas = _hooks.useOas.call(void 0, );
  const operation = _hooks.useOperation.call(void 0, );
  const file = getFile(operation);
  const schemas = getSchemas(operation);
  const generator = new SchemaGenerator(plugin.options, {
    oas,
    plugin,
    pluginManager,
    mode,
    override: plugin.options.override
  });
  const items = [schemas.pathParams, schemas.queryParams, schemas.headerParams, schemas.statusCodes, schemas.request, schemas.response].flat().filter(Boolean);
  const mapItem = ({ name, schema, description, keysToOmit, ...options }, i) => {
    const required = Array.isArray(_optionalChain([schema, 'optionalAccess', _31 => _31.required])) ? !!schema.required.length : !!_optionalChain([schema, 'optionalAccess', _32 => _32.required]);
    const optional = !required && !!name.includes("Params");
    const tree = generator.parse({ schema, name });
    return /* @__PURE__ */ _jsxruntime.jsxs.call(void 0, _components.Oas.Schema, { name, value: schema, tree: [...tree, optional ? { keyword: _pluginoas.schemaKeywords.optional } : void 0].filter(Boolean), children: [
      mode === "split" && /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _components.Oas.Schema.Imports, { extName: plugin.options.extName }),
      /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _react.File.Source, { children: /* @__PURE__ */ _jsxruntime.jsx.call(void 0, OperationSchema, { description, keysToOmit }) })
    ] }, i);
  };
  return /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _react.Parser, { language: "typescript", children: /* @__PURE__ */ _jsxruntime.jsxs.call(void 0, _react.File, { baseName: file.baseName, path: file.path, meta: file.meta, children: [
    /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _react.File.Import, { name: ["z"], path: plugin.options.importPath }),
    items.map(mapItem)
  ] }) });
};

// src/components/Operations.tsx




function Template({ operationsName, pathsName, operations }) {
  const { groupSchemasByName } = _hooks.useOperationManager.call(void 0, );
  const transformedOperations = operations.map((operation) => ({ operation, data: groupSchemasByName(operation, { type: "function" }) }));
  const operationsJSON = transformedOperations.reduce(
    (prev, acc) => {
      prev[`"${acc.operation.getOperationId()}"`] = acc.data;
      return prev;
    },
    {}
  );
  const pathsJSON = transformedOperations.reduce(
    (prev, acc) => {
      prev[`"${acc.operation.path}"`] = {
        ...prev[`"${acc.operation.path}"`] || {},
        [acc.operation.method]: `operations["${acc.operation.getOperationId()}"]`
      };
      return prev;
    },
    {}
  );
  return /* @__PURE__ */ _jsxruntime.jsxs.call(void 0, _jsxruntime.Fragment, { children: [
    /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _react.Const, { export: true, name: operationsName, asConst: true, children: `{${_transformers2.default.stringifyObject(operationsJSON)}}` }),
    /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _react.Const, { export: true, name: pathsName, asConst: true, children: `{${_transformers2.default.stringifyObject(pathsJSON)}}` })
  ] });
}
function RootTemplate({ children }) {
  const {
    mode,
    pluginManager,
    plugin: {
      key: pluginKey,
      options: { extName }
    }
  } = _react.useApp.call(void 0, );
  const { getFile } = _hooks.useOperationManager.call(void 0, );
  const operations = _hooks.useOperations.call(void 0, );
  const { groupSchemasByName } = _hooks.useOperationManager.call(void 0, );
  const transformedOperations = operations.map((operation) => ({ operation, data: groupSchemasByName(operation, { type: "function" }) }));
  const file = pluginManager.getFile({ name: "operations", extName: ".ts", pluginKey });
  const imports = Object.entries(transformedOperations).map(([_key, { data, operation }], index) => {
    const names = [data.request, ...Object.values(data.responses), ...Object.values(data.parameters)].filter(Boolean);
    return /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _react.File.Import, { extName, name: names, root: file.path, path: getFile(operation).path }, index);
  }).filter(Boolean);
  return /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _react.Parser, { language: "typescript", children: /* @__PURE__ */ _jsxruntime.jsxs.call(void 0, _react.File, { baseName: file.baseName, path: file.path, meta: file.meta, exportable: false, children: [
    mode === "split" && imports,
    /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _react.File.Source, { children })
  ] }) });
}
var defaultTemplates = { default: Template, root: RootTemplate };
function Operations({ Template: Template2 = defaultTemplates.default }) {
  const operations = _hooks.useOperations.call(void 0, );
  return /* @__PURE__ */ _jsxruntime.jsx.call(void 0, Template2, { operationsName: "operations", pathsName: "paths", operations });
}
Operations.File = function(props) {
  const templates = { ...defaultTemplates, ...props.templates };
  const Template2 = templates.default;
  const RootTemplate2 = templates.root;
  return /* @__PURE__ */ _jsxruntime.jsx.call(void 0, RootTemplate2, { children: /* @__PURE__ */ _jsxruntime.jsx.call(void 0, Operations, { Template: Template2 }) });
};
Operations.templates = defaultTemplates;

// src/OperationGenerator.tsx

var OperationGenerator = class extends _pluginoas.OperationGenerator {
  async all(operations, _operationsByMethod) {
    const { pluginManager, oas, plugin, mode } = this.context;
    const root = _react.createRoot.call(void 0, {
      logger: pluginManager.logger
    });
    const templates = {
      operations: Operations.templates,
      ...this.options.templates
    };
    root.render(
      /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _react.App, { pluginManager, plugin, mode, children: /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _components.Oas, { oas, operations, generator: this, children: templates.operations && /* @__PURE__ */ _jsxruntime.jsx.call(void 0, Operations.File, { templates: templates.operations }) }) })
    );
    return root.files;
  }
  async operation(operation, options) {
    const { oas, pluginManager, plugin, mode } = this.context;
    const root = _react.createRoot.call(void 0, {
      logger: pluginManager.logger
    });
    root.render(
      /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _react.App, { pluginManager, plugin: { ...plugin, options }, mode, children: /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _components.Oas, { oas, operations: [operation], generator: this, children: /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _components.Oas.Operation, { operation, children: /* @__PURE__ */ _jsxruntime.jsx.call(void 0, OperationSchema.File, {}) }) }) })
    );
    return root.files;
  }
};

// src/plugin.ts
var pluginZodName = "plugin-zod";
var pluginZod = _core.createPlugin.call(void 0, (options) => {
  const {
    output = { path: "zod" },
    group,
    exclude = [],
    include,
    override = [],
    transformers: transformers4 = {},
    dateType = "string",
    unknownType = "any",
    typed = false,
    typedSchema = false,
    mapper = {},
    templates,
    importPath = "zod",
    coercion = false
  } = options;
  const template = _optionalChain([group, 'optionalAccess', _33 => _33.output]) ? group.output : `${output.path}/{{tag}}Controller`;
  return {
    name: pluginZodName,
    options: {
      extName: output.extName,
      transformers: transformers4,
      include,
      exclude,
      override,
      typed,
      typedSchema,
      dateType,
      unknownType,
      mapper,
      importPath,
      coercion,
      templates: {
        operations: Operations.templates,
        ...templates
      }
    },
    pre: [_pluginoas.pluginOasName, typed ? _swaggerts.pluginTsName : void 0].filter(Boolean),
    resolvePath(baseName, pathMode, options2) {
      const root = _path2.default.resolve(this.config.root, this.config.output.path);
      const mode = _nullishCoalesce(pathMode, () => ( _core.FileManager.getMode(_path2.default.resolve(root, output.path))));
      if (mode === "single") {
        return _path2.default.resolve(root, output.path);
      }
      if (_optionalChain([options2, 'optionalAccess', _34 => _34.tag]) && _optionalChain([group, 'optionalAccess', _35 => _35.type]) === "tag") {
        const tag = _transformers.camelCase.call(void 0, options2.tag);
        return _path2.default.resolve(root, _utils.renderTemplate.call(void 0, template, { tag }), baseName);
      }
      return _path2.default.resolve(root, output.path, baseName);
    },
    resolveName(name, type) {
      let resolvedName = _transformers.camelCase.call(void 0, name, {
        suffix: type ? "schema" : void 0,
        isFile: type === "file"
      });
      if (type === "type") {
        resolvedName = _transformers.pascalCase.call(void 0, resolvedName);
      }
      if (type) {
        return _optionalChain([transformers4, 'optionalAccess', _36 => _36.name, 'optionalCall', _37 => _37(resolvedName, type)]) || resolvedName;
      }
      return resolvedName;
    },
    async writeFile(path2, source) {
      if (!path2.endsWith(".ts") || !source) {
        return;
      }
      return this.fileManager.write(path2, source, { sanity: false });
    },
    async buildStart() {
      const [swaggerPlugin] = _core.PluginManager.getDependedPlugins(this.plugins, [_pluginoas.pluginOasName]);
      const oas = await swaggerPlugin.api.getOas();
      const root = _path2.default.resolve(this.config.root, this.config.output.path);
      const mode = _core.FileManager.getMode(_path2.default.resolve(root, output.path));
      const schemaGenerator = new SchemaGenerator(this.plugin.options, {
        oas,
        pluginManager: this.pluginManager,
        plugin: this.plugin,
        contentType: swaggerPlugin.api.contentType,
        include: void 0,
        override,
        mode,
        output: output.path
      });
      const schemaFiles = await schemaGenerator.build();
      await this.addFile(...schemaFiles);
      const operationGenerator = new OperationGenerator(this.plugin.options, {
        oas,
        pluginManager: this.pluginManager,
        plugin: this.plugin,
        contentType: swaggerPlugin.api.contentType,
        exclude,
        include,
        override,
        mode
      });
      const operationFiles = await operationGenerator.build();
      await this.addFile(...operationFiles);
    },
    async buildEnd() {
      if (this.config.output.write === false) {
        return;
      }
      const root = _path2.default.resolve(this.config.root, this.config.output.path);
      if (_optionalChain([group, 'optionalAccess', _38 => _38.type]) === "tag") {
        const rootFiles = await _utils3.getGroupedByTagFiles.call(void 0, {
          logger: this.logger,
          files: this.fileManager.files,
          plugin: this.plugin,
          template,
          exportAs: group.exportAs || "{{tag}}Schemas",
          root,
          output
        });
        await this.addFile(...rootFiles);
      }
      await this.fileManager.addIndexes({
        root,
        output,
        meta: { pluginKey: this.plugin.key },
        logger: this.logger
      });
    }
  };
});

// src/components/Schema.tsx

function Schema(props) {
  const { keysToOmit, withTypeAnnotation, description } = props;
  const { tree, name } = _hooks.useSchema.call(void 0, );
  const {
    pluginManager,
    plugin: {
      options: { mapper, typedSchema, coercion }
    }
  } = _react.useApp.call(void 0, );
  const resolvedName = pluginManager.resolveName({
    name,
    pluginKey: [pluginZodName],
    type: "function"
  });
  const resolvedTypeName = pluginManager.resolveName({
    name,
    pluginKey: [pluginZodName],
    type: "type"
  });
  const typeName = pluginManager.resolveName({
    name,
    pluginKey: [_swaggerts.pluginTsName],
    type: "type"
  });
  if (!tree.length) {
    return /* @__PURE__ */ _jsxruntime.jsx.call(void 0, 
      _react.Const,
      {
        name: resolvedName,
        export: true,
        JSDoc: {
          comments: [description ? `@description ${_transformers2.default.jsStringEscape(description)}` : void 0].filter(Boolean)
        },
        children: "undefined"
      }
    );
  }
  const hasTuple = tree.some((item) => _pluginoas.isKeyword.call(void 0, item, _pluginoas.schemaKeywords.tuple));
  const output = sort(tree).filter((item) => {
    if (hasTuple && (_pluginoas.isKeyword.call(void 0, item, _pluginoas.schemaKeywords.min) || _pluginoas.isKeyword.call(void 0, item, _pluginoas.schemaKeywords.max))) {
      return false;
    }
    return true;
  }).map((item) => parse(void 0, item, { name, typeName, description, mapper, coercion, keysToOmit })).filter(Boolean).join("");
  const suffix = output.endsWith(".nullable()") ? ".unwrap().and" : ".and";
  return /* @__PURE__ */ _jsxruntime.jsxs.call(void 0, _jsxruntime.Fragment, { children: [
    /* @__PURE__ */ _jsxruntime.jsx.call(void 0, 
      _react.Const,
      {
        export: true,
        name: resolvedName,
        JSDoc: {
          comments: [description ? `@description ${_transformers2.default.jsStringEscape(description)}` : void 0].filter(Boolean)
        },
        children: [
          output,
          _optionalChain([keysToOmit, 'optionalAccess', _39 => _39.length]) ? `${suffix}(z.object({ ${keysToOmit.map((key) => `${key}: z.never()`).join(",")} }))` : void 0,
          withTypeAnnotation && typeName ? ` as z.ZodType<${typeName}>` : ""
        ].filter(Boolean).join("") || ""
      }
    ),
    typedSchema && /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _react.Type, { export: true, name: resolvedTypeName, children: `z.infer<typeof ${resolvedName}>` })
  ] });
}
Schema.File = function({}) {
  const {
    pluginManager,
    plugin: {
      options: { typed }
    }
  } = _react.useApp.call(void 0, );
  const { tree, schema } = _hooks.useSchema.call(void 0, );
  const withData = tree.some(
    (schema2) => schema2.keyword === _pluginoas.schemaKeywords.array || schema2.keyword === _pluginoas.schemaKeywords.and || schema2.keyword === _pluginoas.schemaKeywords.object || schema2.keyword === _pluginoas.schemaKeywords.union || schema2.keyword === _pluginoas.schemaKeywords.tuple
  );
  const withTypeAnnotation = !!typed;
  return /* @__PURE__ */ _jsxruntime.jsxs.call(void 0, _components.Oas.Schema.File, { output: pluginManager.config.output.path, children: [
    /* @__PURE__ */ _jsxruntime.jsx.call(void 0, Schema.Imports, {}),
    /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _react.File.Source, { children: /* @__PURE__ */ _jsxruntime.jsx.call(void 0, Schema, { withTypeAnnotation, description: _optionalChain([schema, 'optionalAccess', _40 => _40.description]) }) })
  ] });
};
Schema.Imports = () => {
  const {
    pluginManager,
    plugin: {
      options: { typed, importPath }
    }
  } = _react.useApp.call(void 0, );
  const { path: root } = _react.useFile.call(void 0, );
  const { name, tree, schema } = _hooks.useSchema.call(void 0, );
  const typeName = pluginManager.resolveName({
    name,
    pluginKey: [_swaggerts.pluginTsName],
    type: "type"
  });
  const typeFileName = pluginManager.resolveName({
    name,
    pluginKey: [_swaggerts.pluginTsName],
    type: "file"
  });
  const typePath = pluginManager.resolvePath({
    baseName: typeFileName,
    pluginKey: [_swaggerts.pluginTsName]
  });
  const withTypeAnnotation = !!typed;
  return /* @__PURE__ */ _jsxruntime.jsxs.call(void 0, _jsxruntime.Fragment, { children: [
    /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _react.File.Import, { name: ["z"], path: importPath }),
    withTypeAnnotation && typeName && typePath && /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _react.File.Import, { isTypeOnly: true, root, path: typePath, name: [typeName] })
  ] });
};







exports.Schema = Schema; exports.OperationSchema = OperationSchema; exports.Operations = Operations; exports.pluginZodName = pluginZodName; exports.pluginZod = pluginZod;
